<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Interactive web application for Omics-data analysis • RFLOMICS</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="bootstrap-toc.css">
<script src="bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="pkgdown.css" rel="stylesheet">
<script src="pkgdown.js"></script><meta property="og:title" content="Interactive web application for Omics-data analysis">
<meta property="og:description" content="R-package with shiny interface, provides a framework for the analysis of transcriptomics, proteomics and/or metabolomics data. The interface offers a guided experience for the user, from the definition of the experimental design to the integration of several omics table together. A report can be generated with all settings and analysis results.">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-home">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="index.html">RFLOMICS</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.0.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="articles/RFLOMICS.html">Get started</a>
</li>
<li>
  <a href="reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Manuals
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="articles/RFLOMICS-cmd-line.html">RFLOMICS vignettes cmd line</a>
    </li>
    <li>
      <a href="articles/RFLOMICS-input-data.html">RFLOMICS input data</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://forgemia.inra.fr/flomics/rflomics" class="external-link">
    <span class="fa fa-gitlab"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="contents col-md-9">
<div class="section level1">
<div class="page-header"><h1 id="rflomics-r-package-and-shiny-interface-for-integrative-analysis-of-omics-data">RFLOMICS: R package and Shiny interface for Integrative analysis of omics data<a class="anchor" aria-label="anchor" href="#rflomics-r-package-and-shiny-interface-for-integrative-analysis-of-omics-data"></a>
</h1></div>
<p>The acquisition of <strong>multi-omics data</strong> in the context of <strong>complex experimental design</strong> is a widely used practice to identify features (transcripts, proteins, metabolites,…) and decipher the biological processes they are involved in. The investigation of each omics layer is a good first step to explore and extract relevant biological variability. The statistical integration could then be restrained to pertinent omics levels and features. Such analysis of heterogeneous data remains a technical challenge with the needs of expertise methods and parameters to take into account data specificity. Furthermore, applying different statistical methods from several tools is also a technical challenge in term of data management. In this context, we developed RFLOMICS.</p>
<p><strong>RFLOMICS</strong> is an R package and Shiny interface that offers guided, comprehensive, and reproducible analysis within a framework designed to manage multiple omics datasets. The interface provides remarkable flexibility, enabling users to seamlessly navigate between result exploration and visualization.</p>
<p><img src="reference/figures/workflow.png" align="center" width="600"></p>
<p><strong>Specifications:</strong></p>
<ul>
<li>Performs complete multi-omics project analysis,</li>
<li>Support multi-factorial experimental design (up to 3 biological factors),</li>
<li>Guarantees the relevance of the used methods,</li>
<li>Reduces the analysis time on the unavoidable steps,</li>
<li>Ensure the reproducibility of omics analysis,</li>
<li>Accessible via one and simple user-friendly interface</li>
</ul>
<p><strong>Constraints:</strong></p>
<ul>
<li>3 types of omics data : RNAseq (read count), metabolomics (abundance) and proteomics (abundance)</li>
<li>less than 10 datasets per omics type.</li>
<li>1 to 3 biological factors</li>
<li>1 to 2 batch factors and at least 3 replicates per condition</li>
<li>Complete (mandatory) and balanced (recommended) design</li>
</ul>
<p><strong>Repositories:</strong></p>
<ul>
<li>
<a href="https://forgemia.inra.fr/flomics/rflomics/" class="external-link">forgeMIA</a>: official repository</li>
<li>
<a href="https://github.com/ijpb-bioinformatics/RFLOMICS/" class="external-link">github</a>: mirror repository</li>
</ul>
<div class="section level2">
<h2 id="use-rflomics">Use Rflomics<a class="anchor" aria-label="anchor" href="#use-rflomics"></a>
</h2>
<div class="section level3">
<h3 id="sk8-platform">SK8 platform<a class="anchor" aria-label="anchor" href="#sk8-platform"></a>
</h3>
<p>You can access the rflomics app on Sk8 using this url:</p>
<p><a href="https://rflomics.sk8.inrae.fr/" class="external-link uri">https://rflomics.sk8.inrae.fr/</a></p>
<p>WARNING: only 2Gb of ram available per app, we recommand using the SK8 solution only if your data have a small number of samples and features.</p>
</div>
<div class="section level3">
<h3 id="locally">Locally<a class="anchor" aria-label="anchor" href="#locally"></a>
</h3>
<div class="section level4">
<h4 id="install-from-archive">Install from archive<a class="anchor" aria-label="anchor" href="#install-from-archive"></a>
</h4>
<p>Download the archive from <a href="https://forgemia.inra.fr/flomics/rflomics/-/archive/master/rflomics-master.tar.gz" class="external-link">forgeMIA</a> (tar.gz) or from the <a href="https://github.com/ijpb-bioinformatics/RFLOMICS/archive/refs/heads/master.zip" class="external-link">github mirror</a> (zip)</p>
<p>You can then install the package from the archive using either the install utility from Rstudio or running the following command:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"rflomics-master.tar.gz"</span>, repos <span class="op">=</span> <span class="cn">NULL</span>, type <span class="op">=</span> <span class="st">"source"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="install-from-forgemia-or-github-using-remotes">Install from forgeMIA or github using remotes<a class="anchor" aria-label="anchor" href="#install-from-forgemia-or-github-using-remotes"></a>
</h4>
<p>In a R console, use the following command to install from either the forgeMIA repository or the github repository:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">remotes</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_gitlab.html" class="external-link">install_gitlab</a></span><span class="op">(</span>repo <span class="op">=</span> <span class="st">"flomics/rflomics"</span>, host <span class="op">=</span> <span class="st">"https://forgemia.inra.fr/"</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">remotes</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">"https://github.com/ijpb-bioinformatics/RFLOMICS"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="install-with-git-repository">Install with git repository<a class="anchor" aria-label="anchor" href="#install-with-git-repository"></a>
</h4>
<p>You can clone the repository from forgeMIA using the following command in a shell terminal:</p>
<pre><code>git clone https://forgemia.inra.fr/flomics/rflomics.git</code></pre>
<p>In a R console, set the working directory to the git repository, then install the package:</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/getwd.html" class="external-link">setwd</a></span><span class="op">(</span><span class="st">"rflomics/"</span><span class="op">)</span></span>
<span><span class="fu">remotes</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_local.html" class="external-link">install_local</a></span><span class="op">(</span>upgrade<span class="op">=</span><span class="st">"never"</span><span class="op">)</span></span></code></pre></div>
</div>
</div>
<div class="section level3">
<h3 id="troubleshooting-mofa2">Troubleshooting MOFA2<a class="anchor" aria-label="anchor" href="#troubleshooting-mofa2"></a>
</h3>
<p>Rflomics uses the R package <a href="https://www.bioconductor.org/packages/release/bioc/html/MOFA2.html" class="external-link">MOFA2</a>. This package depends on a python script and this can lead to several issues. The first step is to consult the MOFA2 troubleshooting <a href="https://biofam.github.io/MOFA2/troubleshooting.html" class="external-link">FAQ</a> on their website.</p>
<p>If none of the proposed solutions are resolving your issues, you can try some additional steps and verifications.</p>
<ol style="list-style-type: decimal">
<li>Install <a href="https://www.python.org/downloads/" class="external-link">Python</a> or make sure you have it installed.</li>
</ol>
<p>You can check you have python installed by running <code>which python</code> in a shell terminal. On windows, replace with <code>where python</code>. You should be able to see the path(s) were Python binaries are installed. You can have multiple python version installed at once. You have to make sure the one used by R is this correct one.</p>
<ol start="2" style="list-style-type: decimal">
<li>Install pip command (if not already done). For windows users, you can use <a href="https://phoenixnap.com/kb/install-pip-windows" class="external-link">this guide</a>.</li>
<li>Install mofapy 2: in a shell terminal, run <code>pip install mofapy2</code>.</li>
<li>Set the value of the RETICULATE_PYTHON environment variable to a Python binary and restart your session:</li>
</ol>
<ul>
<li>either manually indicate it in your .Rprofile (hidden file)</li>
<li>or use the following command in a terminal, in the .Rprofile folder:</li>
</ul>
<pre><code>echo "Sys.setenv(RETICULATE_PYTHON = \"path_to_python_bin\")" &gt;&gt; .Rprofile</code></pre>
<ol start="5" style="list-style-type: decimal">
<li>Install MOFA2 using <code>BiocManager::install("MOFA2")</code> in R.</li>
</ol>
<p>You can check your configuration using <code><a href="https://rstudio.github.io/reticulate/reference/py_config.html" class="external-link">reticulate::py_config()</a></code> in R. Your Python binary path and mofapy2 path should appear after step 5. If it’s not the case, something went wrong.</p>
</div>
</div>
<div class="section level2">
<h2 id="run-rflomics">Run Rflomics<a class="anchor" aria-label="anchor" href="#run-rflomics"></a>
</h2>
<p>To run Rflomics, use the runRFLOMICS() function:</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://forgemia.inra.fr/flomics/rflomics/" class="external-link">RFLOMICS</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="reference/runRFLOMICS.html">runRFLOMICS</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="for-additional-information-vignettes">For additional information: <a href="https://flomics.pages.mia.inra.fr/rflomics/index.html" class="external-link">vignettes</a>
<a class="anchor" aria-label="anchor" href="#for-additional-information-vignettes"></a>
</h2>
</div>
<div class="section level2">
<h2 id="licence">Licence<a class="anchor" aria-label="anchor" href="#licence"></a>
</h2>
<p>Artistic-2.0</p>
</div>
<div class="section level2">
<h2 id="contact-and-support">Contact and support<a class="anchor" aria-label="anchor" href="#contact-and-support"></a>
</h2>
<p><a href="mailto:ijpb-bioinfo-team@inrae.fr">ijpb-bioinfo-team</a> (<a href="mailto:ijpb-bioinfo-team@inrae.fr" class="email">ijpb-bioinfo-team@inrae.fr</a>)</p>
</div>
<div class="section level2">
<h2 id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h2>
<ul>
<li>
<a href="https://sysmics.cati.inrae.fr/" class="external-link">CATI Sysmics</a>,</li>
<li><a href="http://eutils.ncbi.nlm.nih.gov/entrez/eutils/elink.fcgi?dbfrom=pubmed&amp;id=32426025&amp;retmode=ref&amp;cmd=prlinks" class="external-link">Ilana L. et al. (2020)</a></li>
<li><a href="">Efstathiou G, et al. 2017</a></li>
</ul>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <div class="links">
<h2 data-toc-skip>Links</h2>
<ul class="list-unstyled">
<li><a href="https://forgemia.inra.fr/flomics/rflomics/-/issues" class="external-link">Report a bug</a></li>
</ul>
</div>

<div class="license">
<h2 data-toc-skip>License</h2>
<ul class="list-unstyled">
<li><a href="https://www.r-project.org/Licenses/Artistic-2.0" class="external-link">Artistic-2.0</a></li>
</ul>
</div>


<div class="citation">
<h2 data-toc-skip>Citation</h2>
<ul class="list-unstyled">
<li><a href="authors.html#citation">Citing RFLOMICS</a></li>
</ul>
</div>

<div class="developers">
<h2 data-toc-skip>Developers</h2>
<ul class="list-unstyled">
<li>Nadia Bessoltane <br><small class="roles"> Author, maintainer </small> <a href="https://orcid.org/0000-0001-6931-2529" target="orcid.widget" aria-label="ORCID" class="external-link"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a> </li>
<li>Delphine Charif <br><small class="roles"> Author </small> <a href="https://orcid.org/0000-0002-1949-5969" target="orcid.widget" aria-label="ORCID" class="external-link"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a> </li>
<li>Audrey Hulot <br><small class="roles"> Author </small> <a href="https://orcid.org/0000-0002-9647-6470" target="orcid.widget" aria-label="ORCID" class="external-link"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a> </li>
<li>Christine Paysant-Leroux <br><small class="roles"> Author </small>  </li>
<li>Gwendal Cueff <br><small class="roles"> Author </small>  </li>
</ul>
</div>



  </div>
</div>


      <footer><div class="copyright">
  <p></p>
<p>Developed by Nadia Bessoltane, Delphine Charif, Audrey Hulot, Christine Paysant-Leroux, Gwendal Cueff.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.9.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
