<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>RFLOMICS • RFLOMICS</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="RFLOMICS">
<meta property="og:description" content="RFLOMICS">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">RFLOMICS</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.0.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../articles/RFLOMICS.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Manuals
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/RFLOMICS-cmd-line.html">RFLOMICS vignettes cmd line</a>
    </li>
    <li>
      <a href="../articles/RFLOMICS-input-data.html">RFLOMICS input data</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://forgemia.inra.fr/flomics/rflomics" class="external-link">
    <span class="fa fa-gitlab"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>RFLOMICS</h1>
            
            <h4 data-toc-skip class="date">July 08, 2024</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/RFLOMICS/RFLOMICS/blob/HEAD/vignettes/RFLOMICS.Rmd" class="external-link"><code>vignettes/RFLOMICS.Rmd</code></a></small>
      <div class="hidden name"><code>RFLOMICS.Rmd</code></div>

    </div>

    
        <div class="abstract">
      <p class="abstract">Abstract</p>
      <p>This document will go through singles and multi-omics data
      analysis with the RFLOMICS package. In a first part, it will be
      shown how to set up the statistical framework of the designed
      protocol: ie translate biological condition comparisons into
      contrasts, perform quality controls, data pre-processing and
      statistical analysis of each omics dataset: ie. find
      differentially expressed genes/proteins/metabolites (DEF) per
      contrasts, identify clusters of co-expressed genes and perform
      functionnal annotation (of type over representation analysis) by
      GO, KEGG or a custom annotation file of the lists of DEF or
      clusters of co-expressed genes/proteins/metabolites. In a second
      part, the vignette will show how to explore common
      variance/correlation between the differents omics datasets and
      simultaneously identify key variables (genes/proteins/metabolites)
      that best correlate to biological factors.</p>
    </div>
    
<div class="section level2">
<h2 id="run-rflomics">Run RFLOMICS<a class="anchor" aria-label="anchor" href="#run-rflomics"></a>
</h2>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/RFLOMICS/RFLOMICS" class="external-link">RFLOMICS</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/runRFLOMICS.html">runRFLOMICS</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p>This vignette will follow the different steps of multi-omics analysis
trough the RFLOMICS interface which are:</p>
<ul>
<li>Data loading</li>
<li>Statistical framework setting</li>
<li>Single omics analysis of each dataset</li>
<li>Single omics results summary</li>
<li>Multi-omics analysis (with the constraint of at least 2
pre-processed datasets)</li>
</ul>
<p>They will appear in the left tab menu of the interface along with the
analysis progression. Whereas single omics analysis steps will be
displayed in tab panel associated to each dataset .</p>
</div>
<div class="section level2">
<h2 id="load-data">Load Data<a class="anchor" aria-label="anchor" href="#load-data"></a>
</h2>
<p>Data loading start by setting a project name (it is mandatory), then
load the experimental design and finally the different omics
datasets.</p>
<div class="section level3">
<h3 id="dataset-example-description">Dataset example description:<a class="anchor" aria-label="anchor" href="#dataset-example-description"></a>
</h3>
<p>Data used for this vignette have been provided by Pr. Loic Rajjou and
Gwendal Cueff. It is included in the ‘ExamplesFiles’ directory of the
package. Briefly, A. thaliana’s transcriptome, metabolome and proteome
have been obtained in the context of the study of seed germination and
vigor. In particular, the authors were interested in deciphering key
entities involved in response to environmental stresses (on the mother
plant): influences of temperature (high, medium and low) and imbibition
stage (Dry: DI, early imbibition: EI and late imbibition: LI). See(<a href="https://www.uibk.ac.at/botany/ecoseed/home/" class="external-link uri">https://www.uibk.ac.at/botany/ecoseed/home/</a>)</p>
</div>
<div class="section level3">
<h3 id="load-experimental-design">Load Experimental Design<a class="anchor" aria-label="anchor" href="#load-experimental-design"></a>
</h3>
<p>The experimental design file must contain experimental
information/conditions for each sample in a tab separated values format
(see RFLOMICS-input-data.html). As soon as this file is loaded, for each
design factor, we have to set up the factor type (Biological factor:
Bio; batch factor: batch; metadata information: meta) and a reference
level.</p>
<ul>
<li><p><em>It is required to have a minimum of 1 biological factor
(maximum 3) and 1 batch factor. Order of the columns does not
matter.</em></p></li>
<li><p><em>It is possible to remove factor’s conditions from the design.
Associated samples will be excluded from the analysis even if they are
present in the omics dataset matrix.</em></p></li>
<li><p><em>It is also possible to re-order factor
modalities</em></p></li>
</ul>
</div>
<div class="section level3">
<h3 id="load-omics-data">Load Omics data<a class="anchor" aria-label="anchor" href="#load-omics-data"></a>
</h3>
<p>Experimental data must be loaded as matrix files in a tab separated
values format. (for more information see RFLOMICS-input-data.html). For
each dataset it is mandatory to specify the type of omics
(transcriptomics, proteomics, metabolomics ) as it will adapt the data
processing and analysis methods. By default, datasets will be called
set1, set2, set3, …, respectively. A more suitable name can be given by
changing ‘Dataset name’ in the interface.</p>
<p>To load the 3 examples datasets of the Ecoseed project: RNAseq data
(read counts), proteomics data (abundance of proteins), and metabolomics
data (intensity of metabolites) (see RFLOMICS-input-data.html), please
press the ‘Load Ecoseed Data’.</p>
<p>As soon as the data are loaded the <strong>RFlomics
MultiAssayExperiment</strong> object is created. The data overview panel
displays the samples overlap across datasets and the number of datasets
per combination of biological factors.</p>
<p><img src="Images/01_loadData.png" style="width:80.0%"></p>
</div>
</div>
<div class="section level2">
<h2 id="set-the-statistical-framework">Set the statistical framework<a class="anchor" aria-label="anchor" href="#set-the-statistical-framework"></a>
</h2>
<p>The interface provides models written from the simplest one to the
most complete one. Only the second order interaction terms between the
biological factors appear. Batch factors never appear in interaction
terms. In our example, the complete model is chosen.</p>
<p>All ‘a priori’ hypotheses (contrasts) are then calculated from the
chosen model and displayed. The user must select those that correspond
to the biological questions. Contrasts are numbered. These labels will
follow for all analysis.</p>
<p>In our example we are studying the effect of temperature on seed
germination (transition from the state of dormancy to the state of
germination vigor). For that we select 3 average contrasts (see
image):</p>
<p>This statistical framework will be applied on all loaded
datasets.</p>
<p><img src="Images/02_setModel.png" style="width:80.0%"></p>
</div>
<div class="section level2">
<h2 id="single-omics-data-analysis">Single omics data analysis<a class="anchor" aria-label="anchor" href="#single-omics-data-analysis"></a>
</h2>
<p>Once the contrasts are chosen, 3 Item menus appear in the side bar
menu that correspond to the 3 loaded omics types.</p>
<p>For each dataset, different steps of data analysis are proposed as a
tab panel. These analysis steps must be performed sequentially. It is
possible to switch between datasets. However, you cannot perform a new
task until the previous one is completed. A progress bar is displayed
indicating the progress status.</p>
<div class="section level3">
<h3 id="data-processing-and-quality-check">Data processing and quality check<a class="anchor" aria-label="anchor" href="#data-processing-and-quality-check"></a>
</h3>
<p>In this tab panel we access a data quality control screen and we
perform the appropriate data filtering and processing according to the
type of omics.</p>
<p>By default, features with 0 count in all conditions are removed from
the data.</p>
<p>Specifically for each data we have the possibility to exclude samples
or to select a subset of samples to analyze, provided the completeness
condition is met.</p>
<div class="section level4">
<h4 id="completeness-check">Completeness check<a class="anchor" aria-label="anchor" href="#completeness-check"></a>
</h4>
<p>The design must be complete and as possible balanced. A
<strong>Complete design (mandatory)</strong> mean that all possible
combinations of biological factors’ levels (called groups) are presents.
A <strong>Balanced design (recommended)</strong> has an equal number of
o bservations (replicates) for each group.</p>
<p>These constraints are checked and represented as graph (each square
represent 1 group and each color level indicate the number of
observations).</p>
<p><img src="Images/00_completeDesign.png" width="40%" style="display: block; margin: auto;"><img src="Images/00_nonBalancedDesign.png" width="40%" style="display: block; margin: auto;"></p>
<p>The effect of data filtering or sample removing step can be explored
thanks to many QC graphs. For each QC, a graph is built on each raw data
and filtered data to allow the comparison.</p>
</div>
<div class="section level4">
<h4 id="transcriptomics-data">Transcriptomics data<a class="anchor" aria-label="anchor" href="#transcriptomics-data"></a>
</h4>
<p>For RNAseq data, low expressed genes are filtered so that only genes
whose cpm expression is greater than <strong>“CPM cutoff”</strong> in x
samples are kept. The value of x is given by <strong>“Filtering
strategy”</strong> (NbOfsample_over_cpm &lt;= NbConditions). By default
the data are filtered based on NbCondition strategy with CPM cutoff
equal to 5. In this example, there are 9 conditions. So, the number of
genes with a CPM less than 5 in at least 9 samples are removed.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># geneToKeep &lt;- counts[rowSums(edgeR::cpm(counts) &gt;= CPM_Cutoff) &gt;= x, ]</span></span></code></pre></div>
<p>To correct for differences in sequencing depths (library size),
Rflomics proposes one method for normalization, <strong>“TMM
method”</strong> from edgeR package.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># dge &lt;- edgeR::DGEList(counts=counts, group=groups)</span></span>
<span><span class="co"># dge &lt;- edgeR::calcNormFactors(dge, method="TMM")</span></span></code></pre></div>
<p>For RNAseq data the QC graphs are drawn before and after the
filtering and normalization steps. The following graphs are
displayed:</p>
<ul>
<li>The log2(count) distribution per sample: Expect aligned medians.
Samples with shifted median may be outliers.</li>
<li>The library size distribution which is the distribution of the total
number of reads per sample.</li>
<li>The density distribution of log2(count) per sample: Expect a
unimodal gaussian density distribution.</li>
<li>An interactive Principal Component Analysis allowing to color
samples by any factor of the design to decode the main sources of
variation in the data.</li>
</ul>
<p><img src="Images/03_DataProcess.png" style="width:80.0%"></p>
</div>
</div>
<div class="section level3">
<h3 id="differential-expression-analysis">Differential expression analysis<a class="anchor" aria-label="anchor" href="#differential-expression-analysis"></a>
</h3>
<p>In this tab panel, we can run the differential expression analysis.
The analysis is run by contrast.<br>
For RNAseq data, the <strong>glmFit()</strong> function from the
<strong>edgeR</strong> package is used whereas for proteomic and
metabolomic data, the <strong>lmFit()</strong> function from
<strong>limma</strong> packages is used. The ability to run analysis on
a cluster is offered by sliding the cluster button (see
RFLOMICS-configure-cluster-access vignette)</p>
<p>Results will appear in two components: a scrolling menu will give the
results of the analysis by contrat and, at the bottom of the panel, the
intersecting sets of the lists of DEG is given thanks to the
<strong>UpSetR()</strong> function.</p>
<p>Each results has to be validated:</p>
<ul>
<li>
<p>You <strong>have to</strong> to check the Pvalue distribution for
each contrast and validate it (<strong>ok checkbox</strong>).</p>
<ul>
<li>If the distribution is uniform: it is ok. A few genes are DE and the
FDR will find them.</li>
<li>If the distribution as a peak near 0 and then is uniform: it is also
ok. The higher the peak, the higher the number of DE genes. FDR
correction can be applied.</li>
<li>If the distribution has 2 peaks one near 0 and one near 1 or just
one peak near 1, it is not a good news! You have to understand what’s
happened.</li>
</ul>
</li>
<li><p>You can filter differentially expressed
genes/proteins/metabolites either by fold change (<strong>|FC|</strong>)
or by False Discovery Rate (BH) cut-off. Graphs will be automatically
updated.</p></li>
<li><p>You have to <strong>validate</strong> to save the cut-off
thresholds but also the contrasts result selection and pass to the
co-expression analysis or to the annotation enrichment step.</p></li>
</ul>
<p><img src="Images/04_diffAnal.png" style="width:80.0%"></p>
</div>
<div class="section level3">
<h3 id="coexpression-analysis">Coexpression analysis<a class="anchor" aria-label="anchor" href="#coexpression-analysis"></a>
</h3>
<ul>
<li><p>In this tab panel, choice is given to merge
(<strong>union</strong>) or intersect<br>
(<strong>intersection</strong>) the lists of differentially expressed
genes/proteins/metabolites associated to contrasts.</p></li>
<li>
<p>Co-expression analysis is done using to the
<strong>coseq</strong> package with a set of optimal parameters.</p>
<ul>
<li>For <strong>RNAseq</strong>, data transformation method is set to
the <strong>arcsin</strong> function and normalization method to
<strong>TMM</strong>. No scaling is done for these data.
<strong>gaussian mixture model</strong> is used to decipher the
different profiles of gene expression with a set of parameters to
estimate fixed to (Gaussian_pK_Lk_Ck).</li>
<li>For <strong>proteomics/metabolomics</strong> data, scaling is done
on proteins/metabolites. Another sets of parameters estimation is also
proposed for the modeling of gene expression profile:
(Gaussian_pK_Lk_Bk). It has to be used, if the other one doesn’t
fit.</li>
</ul>
</li>
</ul>
<p>The number of technical replicates to perform (<strong>iter</strong>)
for each <strong>K</strong> which is the number of expressions’ profile
into which the DEG have to be cut, can be first set to 2 to precise the
range of K and then put to a minimum of 20 replicates per K.</p>
<p>The ability to run the clustering analysis on a cluster is offered by
sliding the Cluster button (see Rflomics_configure-cluster-access).</p>
<ul>
<li>
<p>Results description (see <a href="https://www.bioconductor.org/packages/release/bioc/vignettes/coseq/inst/doc/coseq.html" class="external-link">CoSeq</a>)</p>
<ul>
<li><p>Results can be explored with several plots proposed by
<strong>coseq</strong>. The ICL graph has been slightly modified to show
all the range of the ICL values for a given K.</p></li>
<li><p>A table of jobs summary is also given. It groups error messages
by K.</p></li>
</ul>
</li>
</ul>
<p><img src="Images/05_coExpAnal.png" style="width:80.0%"></p>
</div>
<div class="section level3">
<h3 id="annotation-and-enrichment-analysis-using-clusterprofilr">Annotation and Enrichment analysis using clusterProfilR<a class="anchor" aria-label="anchor" href="#annotation-and-enrichment-analysis-using-clusterprofilr"></a>
</h3>
<p>We select the lists of DE genes or clusters to annotate. All
available lists are selected by default.</p>
<p>The enrichment analysis is performed using the <a href="https://www.sciencedirect.com/science/article/pii/S2666675821000667" class="external-link">clusterProfiler
R-package</a>. For more information on the methods and the package,
please see <a href="https://www.sciencedirect.com/science/article/pii/S2666675821000667" class="external-link">The
article</a>, <a href="https://bioconductor.org/packages/release/bioc/html/clusterProfiler.html" class="external-link">Bioconductor
Package</a> and the <a href="https://yulab-smu.top/biomedical-knowledge-mining-book/" class="external-link">clusterProfiler
Vignette</a>.</p>
<p>You can chose between several domains, depending on several
parameters:</p>
<ul>
<li>custom: preferred option. It requires an annotation file (tabulated,
at least two columns, one for the name of the gene, the other for the
name or id of the annotation term). A second panel will allow the user
to indicate which columns indicates what. You can have as many
domains/ontology as you want, you just need to indicate the columns name
that differentiate them in the right area;</li>
<li>GO (or <a href="GO:BP" class="uri">GO:BP</a>, <a href="GO:CC" class="uri">GO:CC</a>, <a href="GO:MF" class="uri">GO:MF</a> if only one
GO domain is of interest). The user will have to chose a database
library (<a href="https://www.bioconductor.org/packages/release/BiocViews.html#___OrgDb" class="external-link">org.*.db</a>)
installed on the libPath and the keytype of the entities (identifier to
make the connection with the annotation).</li>
<li>KEGG: In the setting panel, enter the three letters to identify the
organism and the keytype of the entities. It requires to have an online
connection.</li>
</ul>
<p>Once everything is set and the analysis is conducted, several panels
are displayed to show the results.</p>
<div class="float">
<img src="Images/06_ORAAnal.png" style="width:80.0%" alt="Results of the analysis for multiple lists"><div class="figcaption">Results of the analysis for multiple lists</div>
</div>
</div>
<div class="section level3">
<h3 id="dataset-analysis-summary">Dataset analysis summary<a class="anchor" aria-label="anchor" href="#dataset-analysis-summary"></a>
</h3>
<p>This panel summarizes the effects of the filtering steps on the
selected datasets and the results of the differential analyses. The
first row shows how many samples and entities are left in each table.
The integration will be performed on these datasets. The second row
shows the number of DE entities for each selected contrast and for each
table, with the distinction between up-regulated and down-regulated
entities.</p>
</div>
</div>
<div class="section level2">
<h2 id="integration">Integration<a class="anchor" aria-label="anchor" href="#integration"></a>
</h2>
<p>Once at least two tables have passed the pre-processing step, the
Integration tab will appear in the side bar menu. Two sub menus are
available. For each of them, you will have to run a preprocessing before
accessing to the integration panel: select the datasets you want to
integrate together and the filtering strategy for each data. A filtering
is strongly advised when having large data or unbalanced number of
features between them with the goals to homogenize the number of
features between the datasets.</p>
<p>You can filter according to the DE entities found in the differential
analysis, or based on the variation coefficient, taking a certain number
of variables that have the largest coefficient. For this example, we
chose to keep only 500 features from the RNAseq data, based on the
variation coefficient.</p>
<p>RNAseq data will be transformed before the integration. This is done
with the <strong>voom</strong> transformation from the
<strong>limma</strong> R package.</p>
<p>Batch effects are removed before applying the function, using
<strong>removebatcheffect</strong> from <strong>limma</strong>.</p>
<p>Features are then centered before the integration.</p>
<p><img src="Images/07_prepData.png" style="width:100.0%"></p>
<div class="section level3">
<h3 id="mofa">MOFA<a class="anchor" aria-label="anchor" href="#mofa"></a>
</h3>
<p>This integration analysis is performed using the
<strong>MOFA2</strong> R-package. It is an unsupervised method that
searches for latent factors to decompose the variance of the data,
similarly to a PCA, but applied to a multi-dataset situation. The
interpretation of the results is also very similar to those of a PCA:
the most interesting factors are the one explaining more variance,
whether it’s on several dataset or just one table. It also gives weights
to the features of each table, for each factor. This is interesting in
case a factor can be <strong>a posteriori</strong> linked to a covariate
(in the case displayed on the figure, the temperature seems linked to
the first factor, the features used to build this factor are to be
considered very relevant in the analysis).</p>
<p><img src="Images/07_MOFA.png" style="width:80.0%"></p>
<p>The user can chose to scale or not each tables, with the scale views
argument, then set the number of factors to be computed and finally the
maximum iterations for the MOFA run. We recommend that the user do not
change the number of iterations. Once everything is set, you can run the
analysis by clicking on the run button.</p>
<p>Several figure are provided by the interface once the analysis is
done.</p>
</div>
<div class="section level3">
<h3 id="mixomics">MixOmics<a class="anchor" aria-label="anchor" href="#mixomics"></a>
</h3>
<p>This integration analysis is performed using the
<strong>MixOmics</strong> R-package. In RFLOMICS, we focused on the
supervised analyses using the functions block.(s)plda (also called
DIABLO) from the package. Only complete cases will be used for the
analysis. The user can select all biological factors as response
variables. The meta and batch factors are not available.</p>
<p>Using Sparse Analysis will set the “s” in the function names. It this
case, the tuning ‘cases’ argument will be considered and passed to
tune.splsda or tune.block.splsda: it determines the number of features
selection to be tested.</p>
<p>In RFLOMICS, we chose to only ask for a number of cases. This number
of cases is used to determine the keep.X argument, using a span of
nfeatures/number of cases. By example, if the user chose 5 cases and has
2 datasets with respectively 500 and 250 features, the function will
test 25 combinaisons of features per component. For each datasets,
features will be picked-up in a cumulative manner ie for the DS1:
100,200,300,400,500 and DS2: 50, 100, 150, 200, 250. An example of
combination will be c(DS1=100,DS2=50). The best combination of results
by components will be used by mixOmics block.splsda.</p>
<p>Once everything is set, you can run the analysis by clicking on the
run button. Several figures are provided by the interface once the
analysis is done, for each selected response variable.</p>
<p><img src="Images/07_mixOmics.png" style="width:100.0%"></p>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Nadia Bessoltane, Delphine Charif, Audrey Hulot, Christine Paysant-Leroux, Gwendal Cueff.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.9.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
