<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>RFLOMICS input format • RFLOMICS</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="RFLOMICS input format">
<meta property="og:description" content="RFLOMICS">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">RFLOMICS</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.0.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../articles/RFLOMICS.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Manuals
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/RFLOMICS-cmd-line.html">RFLOMICS vignettes cmd line</a>
    </li>
    <li>
      <a href="../articles/RFLOMICS-input-data.html">RFLOMICS input data</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://forgemia.inra.fr/flomics/rflomics" class="external-link">
    <span class="fa fa-gitlab"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><link href="RFLOMICS-input-data_files/htmltools-fill-0.5.8.1/fill.css" rel="stylesheet">
<script src="RFLOMICS-input-data_files/htmlwidgets-1.6.4/htmlwidgets.js"></script><link href="RFLOMICS-input-data_files/datatables-css-0.0.0/datatables-crosstalk.css" rel="stylesheet">
<script src="RFLOMICS-input-data_files/datatables-binding-0.33/datatables.js"></script><link href="RFLOMICS-input-data_files/dt-core-1.13.6/css/jquery.dataTables.min.css" rel="stylesheet">
<link href="RFLOMICS-input-data_files/dt-core-1.13.6/css/jquery.dataTables.extra.css" rel="stylesheet">
<script src="RFLOMICS-input-data_files/dt-core-1.13.6/js/jquery.dataTables.min.js"></script><link href="RFLOMICS-input-data_files/crosstalk-1.2.1/css/crosstalk.min.css" rel="stylesheet">
<script src="RFLOMICS-input-data_files/crosstalk-1.2.1/js/crosstalk.min.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>RFLOMICS input format</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/RFLOMICS/RFLOMICS/blob/HEAD/vignettes/RFLOMICS-input-data.Rmd" class="external-link"><code>vignettes/RFLOMICS-input-data.Rmd</code></a></small>
      <div class="hidden name"><code>RFLOMICS-input-data.Rmd</code></div>

    </div>

    
    
<p>RFLOMICS can handle three types of omics (RNAseq, proteomics and
metabolomics) and several datasets per omics generated from similar
experimental design. Each dataset must have a complete design.</p>
<p>Below each input file is detailed :</p>
<div class="section level2">
<h2 id="experimental-design-file">Experimental design file<a class="anchor" aria-label="anchor" href="#experimental-design-file"></a>
</h2>
<!-- ce fichier contient le plan experimental (conditions) plus des metadata par echantillons -->
<p>This file provides the experimental design table. The first column
indicates the sample names. Next columns indicate experimental
conditions (called biological factors or technical/batch factors) and
metadata (optional). Each row describes a sample by specifying the level
for each experimental condition. This file is edited by the user and
must contain a header with column names corresponding respectively to
the factor names and metadata names.</p>
<p>It is advised that each factor’s level starts with a letter (for
example, for a factor called Month, it is better to write M1 rather than
1).</p>
<div class="datatables html-widget html-fill-item" id="htmlwidget-ac96cb3ee4656e2e9ec3" style="width:100%;height:auto;"></div>
<script type="application/json" data-for="htmlwidget-ac96cb3ee4656e2e9ec3">{"x":{"filter":"none","vertical":false,"data":[["Elevated_LI_1","Elevated_LI_3","Elevated_LI_2","Elevated_EI_1","Elevated_EI_2","Elevated_EI_3","Elevated_DS_1","Elevated_DS_3","Elevated_DS_2","Low_LI_1","Low_LI_2","Low_LI_3","Low_EI_3","Low_EI_1","Low_EI_2","Low_DS_1","Low_DS_3","Low_DS_2","Medium_LI_1","Medium_LI_2","Medium_LI_3","Medium_EI_2","Medium_EI_1","Medium_EI_3","Medium_DS_1","Medium_DS_3","Medium_DS_2"],["rep1","rep3","rep2","rep1","rep2","rep3","rep1","rep3","rep2","rep1","rep2","rep3","rep3","rep1","rep2","rep1","rep3","rep2","rep1","rep2","rep3","rep2","rep1","rep3","rep1","rep3","rep2"],["Elevated","Elevated","Elevated","Elevated","Elevated","Elevated","Elevated","Elevated","Elevated","Low","Low","Low","Low","Low","Low","Low","Low","Low","Medium","Medium","Medium","Medium","Medium","Medium","Medium","Medium","Medium"],["LI","LI","LI","EI","EI","EI","DS","DS","DS","LI","LI","LI","EI","EI","EI","DS","DS","DS","LI","LI","LI","EI","EI","EI","DS","DS","DS"]],"container":"<table class=\"display\">\n  <thead>\n    <tr>\n      <th> <\/th>\n      <th>Repeat<\/th>\n      <th>temperature<\/th>\n      <th>imbibition<\/th>\n    <\/tr>\n  <\/thead>\n<\/table>","options":{"columnDefs":[{"orderable":false,"targets":0},{"name":" ","targets":0},{"name":"Repeat","targets":1},{"name":"temperature","targets":2},{"name":"imbibition","targets":3}],"order":[],"autoWidth":false,"orderClasses":false},"selection":{"mode":"multiple","selected":null,"target":"row","selectable":null}},"evals":[],"jsHooks":[]}</script>
</div>
<div class="section level2">
<h2 id="omic-data-file">Omic data file<a class="anchor" aria-label="anchor" href="#omic-data-file"></a>
</h2>
<p>For each omics: The matrix-like file with abundance of omics data is
needed. The first column indicates feature names (genes, proteins, or
metabolites), column headers indicate the sample names.</p>
<div class="section level3">
<h3 id="transcriptomics-rnaseq-counts-data">
<strong>Transcriptomics (RNAseq counts data)</strong><a class="anchor" aria-label="anchor" href="#transcriptomics-rnaseq-counts-data"></a>
</h3>
<!-- les comptage de reads sont obtenu après une analyse bioinfo (mapping sur le genome de ref et comptage des reads qui couvrent chaque gene) -->
<p>For RNAseq data the values correspond to raw read counts per genes or
transcripts.</p>
<div class="datatables html-widget html-fill-item" id="htmlwidget-e5c8c404fe174e4c81bd" style="width:100%;height:auto;"></div>
<script type="application/json" data-for="htmlwidget-e5c8c404fe174e4c81bd">{"x":{"filter":"none","vertical":false,"data":[["AT1G01010","AT1G01020","AT1G01030","AT1G01040","AT1G01050","AT1G01060","AT1G01070","AT1G01073","AT1G01080","AT1G01090"],[31,58,4,141,330,18,5,0,26,44],[131,132,35,629,650,58,2,0,61,227],[75,57,17,246,355,24,2,0,35,129],[104,97,31,329,463,28,2,0,53,132],[83,83,31,370,401,16,1,0,52,108]],"container":"<table class=\"display\">\n  <thead>\n    <tr>\n      <th> <\/th>\n      <th>Low_DS_1<\/th>\n      <th>Low_DS_2<\/th>\n      <th>Low_DS_3<\/th>\n      <th>Medium_DS_1<\/th>\n      <th>Medium_DS_2<\/th>\n    <\/tr>\n  <\/thead>\n<\/table>","options":{"columnDefs":[{"className":"dt-right","targets":[1,2,3,4,5]},{"orderable":false,"targets":0},{"name":" ","targets":0},{"name":"Low_DS_1","targets":1},{"name":"Low_DS_2","targets":2},{"name":"Low_DS_3","targets":3},{"name":"Medium_DS_1","targets":4},{"name":"Medium_DS_2","targets":5}],"order":[],"autoWidth":false,"orderClasses":false},"selection":{"mode":"multiple","selected":null,"target":"row","selectable":null}},"evals":[],"jsHooks":[]}</script>
</div>
<div class="section level3">
<h3 id="proteomic-data">
<strong>Proteomic data</strong><a class="anchor" aria-label="anchor" href="#proteomic-data"></a>
</h3>
<p>For proteomics data the values correspond to intensity of
proteins.</p>
<div class="datatables html-widget html-fill-item" id="htmlwidget-36aa3d2a04d42bbc2145" style="width:100%;height:auto;"></div>
<script type="application/json" data-for="htmlwidget-36aa3d2a04d42bbc2145">{"x":{"filter":"none","vertical":false,"data":[["AT1G01900","AT1G02305","AT1G02700","AT1G02780","AT1G03680","AT1G03890","AT1G04170","AT1G04480","AT1G04580","AT1G05510"],[8.348679729200001,6.5452781421,6.9216726893,7.8256109866,7.2988639955,9.0530189041,7.182533637,8.478609007799999,6.9754899786,8.3051859212],[8.6061943001,6.6544128724,6.7546506091,7.6376143383,7.0896049533,9.0751488737,7.0889374085,8.5155982316,7.151373158,8.1587581416],[8.526607883100001,6.9058717324,6.7989753152,7.7438195193,7.0762258389,9.021445458100001,7.1491077231,8.2942954683,7.0233718753,8.092068364399999],[8.420953533,6.7216782338,6.9653282971,7.6477754944,7.2286353235,9.0609187566,7.0256422928,8.224319039799999,6.8504988534,8.4061845606],[8.5519694987,6.6288127033,6.8448249449,7.4977747461,7.2764162093,9.0759374174,7.0655168446,8.146744285600001,7.0795080769,8.4218343557]],"container":"<table class=\"display\">\n  <thead>\n    <tr>\n      <th> <\/th>\n      <th>Elevated_LI_1<\/th>\n      <th>Elevated_LI_3<\/th>\n      <th>Elevated_LI_2<\/th>\n      <th>Elevated_EI_1<\/th>\n      <th>Elevated_EI_2<\/th>\n    <\/tr>\n  <\/thead>\n<\/table>","options":{"columnDefs":[{"className":"dt-right","targets":[1,2,3,4,5]},{"orderable":false,"targets":0},{"name":" ","targets":0},{"name":"Elevated_LI_1","targets":1},{"name":"Elevated_LI_3","targets":2},{"name":"Elevated_LI_2","targets":3},{"name":"Elevated_EI_1","targets":4},{"name":"Elevated_EI_2","targets":5}],"order":[],"autoWidth":false,"orderClasses":false},"selection":{"mode":"multiple","selected":null,"target":"row","selectable":null}},"evals":[],"jsHooks":[]}</script>
</div>
<div class="section level3">
<h3 id="metabolomic-data">
<strong>Metabolomic data</strong><a class="anchor" aria-label="anchor" href="#metabolomic-data"></a>
</h3>
<p>For metabolomics data the values correspond to intensity of
metabolites.</p>
<div class="datatables html-widget html-fill-item" id="htmlwidget-febe03efa1a2d8d52a86" style="width:100%;height:auto;"></div>
<script type="application/json" data-for="htmlwidget-febe03efa1a2d8d52a86">{"x":{"filter":"none","vertical":false,"data":[["m_Alanine","m_Arginine","m_Asparagine","m_Aspartate","m_CystEIn","m_GABA","m_Glutamine","m_Glutamate","m_Glycine","m_Isoleucine"],[0.767047071,0.029325117,0.419735725,0.147360459,0.073655289,0.067871343,0.085097151,0.585539888,0.107736669,0.030546914],[0.7360740609999999,0.034914563,0.444754603,0.149486978,0.06662491399999999,0.06902301299999999,0.111321338,0.593996301,0.116613047,0.029564253],[0.812803189,0.046429932,0.434730864,0.140378787,0.127331686,0.076388119,0.110210481,0.522253647,0.147028681,0.033741244],[0.429987861,0.021028136,0.583022003,0.28173995,0.054387281,0.08047530999999999,0.083014858,1.286398062,0.104911126,0.053982475],[0.513079988,0.011900091,0.545867518,0.308790207,0.033108516,0.070034556,0.126291681,1.174695734,0.118575382,0.057259034]],"container":"<table class=\"display\">\n  <thead>\n    <tr>\n      <th> <\/th>\n      <th>Low_DS_1<\/th>\n      <th>Low_DS_2<\/th>\n      <th>Low_DS_3<\/th>\n      <th>Medium_DS_1<\/th>\n      <th>Medium_DS_2<\/th>\n    <\/tr>\n  <\/thead>\n<\/table>","options":{"columnDefs":[{"className":"dt-right","targets":[1,2,3,4,5]},{"orderable":false,"targets":0},{"name":" ","targets":0},{"name":"Low_DS_1","targets":1},{"name":"Low_DS_2","targets":2},{"name":"Low_DS_3","targets":3},{"name":"Medium_DS_1","targets":4},{"name":"Medium_DS_2","targets":5}],"order":[],"autoWidth":false,"orderClasses":false},"selection":{"mode":"multiple","selected":null,"target":"row","selectable":null}},"evals":[],"jsHooks":[]}</script>
</div>
</div>
<div class="section level2">
<h2 id="annotation-of-features-optional">Annotation of features (optional)<a class="anchor" aria-label="anchor" href="#annotation-of-features-optional"></a>
</h2>
<p>This file contains annotation about biological functions of
genes/proteins/metabolites, or their implication in biological pathways.
This annotation is needed to compute Over Representation Analysis (ORA).
This file must contain at least 2 columns, names of features (identical
to the ones used in the abundance matrix) and term identifiers (ex. GO
term accession : <a href="GO:0034599" class="uri">GO:0034599</a>). It is
possible to add 2 more columns: names of terms (ex. cellular response to
oxidative stress) and the domain of annotation (ex.
biological_process).</p>
<table class="table">
<colgroup>
<col width="16%">
<col width="19%">
<col width="43%">
<col width="20%">
</colgroup>
<thead><tr class="header">
<th>Gene ID</th>
<th>GO term accession</th>
<th>GO term name</th>
<th>GO domain</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>AT4G36648</td>
<td><a href="GO:0006412" class="uri">GO:0006412</a></td>
<td>translation</td>
<td>biological_process</td>
</tr>
<tr class="even">
<td>AT4G36648</td>
<td><a href="GO:0030533" class="uri">GO:0030533</a></td>
<td>triplet codon-amino acid adaptor activity</td>
<td>molecular_function</td>
</tr>
<tr class="odd">
<td>AT1G18745</td>
<td><a href="GO:0006396" class="uri">GO:0006396</a></td>
<td>RNA processing</td>
<td>biological_process</td>
</tr>
<tr class="even">
<td>AT1G18745</td>
<td><a href="GO:0005730" class="uri">GO:0005730</a></td>
<td>nucleolus</td>
<td>cellular_component</td>
</tr>
<tr class="odd">
<td>AT3G00980</td>
<td><a href="GO:0006396" class="uri">GO:0006396</a></td>
<td>RNA processing</td>
<td>biological_process</td>
</tr>
<tr class="even">
<td>AT3G00980</td>
<td><a href="GO:0005730" class="uri">GO:0005730</a></td>
<td>nucleolus</td>
<td>cellular_component</td>
</tr>
</tbody>
</table>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Nadia Bessoltane, Delphine Charif, Audrey Hulot, Christine Paysant-Leroux, Gwendal Cueff.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.9.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
