<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Run Gene Enrichment Analysis and process results — runAnnotationEnrichment • RFLOMICS</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Run Gene Enrichment Analysis and process results — runAnnotationEnrichment"><meta property="og:description" content="This function performs overrepresentation analysis (ORA) using
clusterprofiler functions. It can be used with custom annotation file
(via enricher), GO (enrichGO) or KEGG (enrichKEGG) annotations."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">RFLOMICS</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.0.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../articles/RFLOMICS.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Manuals
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/RFLOMICS-cmd-line.html">RFLOMICS vignettes cmd line</a>
    </li>
    <li>
      <a href="../articles/RFLOMICS-input-data.html">RFLOMICS input data</a>
    </li>
  </ul></li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://forgemia.inra.fr/flomics/rflomics" class="external-link">
    <span class="fa fa-gitlab"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Run Gene Enrichment Analysis and process results</h1>
    
    <div class="hidden name"><code>runAnnotationEnrichment.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>This function performs overrepresentation analysis (ORA) using
clusterprofiler functions. It can be used with custom annotation file
(via enricher), GO (enrichGO) or KEGG (enrichKEGG) annotations.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="co"># S4 method for RflomicsSE</span></span>
<span><span class="fu">runAnnotationEnrichment</span><span class="op">(</span></span>
<span>  <span class="va">object</span>,</span>
<span>  nameList <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  list_args <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>pvalueCutoff <span class="op">=</span> <span class="fl">0.05</span>, qvalueCutoff <span class="op">=</span> <span class="fl">1</span>, minGSSize <span class="op">=</span> <span class="fl">10</span>, maxGSSize <span class="op">=</span></span>
<span>    <span class="fl">500</span>, universe <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">object</span><span class="op">)</span><span class="op">)</span>,</span>
<span>  from <span class="op">=</span> <span class="st">"DiffExp"</span>,</span>
<span>  database <span class="op">=</span> <span class="st">"custom"</span>,</span>
<span>  domain <span class="op">=</span> <span class="st">"no-domain"</span>,</span>
<span>  col_term <span class="op">=</span> <span class="st">"term"</span>,</span>
<span>  col_gene <span class="op">=</span> <span class="st">"gene"</span>,</span>
<span>  col_name <span class="op">=</span> <span class="st">"name"</span>,</span>
<span>  col_domain <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  annot <span class="op">=</span> <span class="cn">NULL</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># S4 method for RflomicsMAE</span></span>
<span><span class="fu">runAnnotationEnrichment</span><span class="op">(</span></span>
<span>  <span class="va">object</span>,</span>
<span>  <span class="va">SE.name</span>,</span>
<span>  nameList <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  list_args <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>pvalueCutoff <span class="op">=</span> <span class="fl">0.05</span>, qvalueCutoff <span class="op">=</span> <span class="fl">1</span>, minGSSize <span class="op">=</span> <span class="fl">10</span>, maxGSSize <span class="op">=</span></span>
<span>    <span class="fl">500</span>, universe <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">object</span><span class="op">)</span><span class="op">)</span>,</span>
<span>  from <span class="op">=</span> <span class="st">"DiffExp"</span>,</span>
<span>  database <span class="op">=</span> <span class="st">"custom"</span>,</span>
<span>  domain <span class="op">=</span> <span class="st">"no-domain"</span>,</span>
<span>  col_term <span class="op">=</span> <span class="st">"term"</span>,</span>
<span>  col_gene <span class="op">=</span> <span class="st">"gene"</span>,</span>
<span>  col_name <span class="op">=</span> <span class="st">"name"</span>,</span>
<span>  col_domain <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  annot <span class="op">=</span> <span class="cn">NULL</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># S4 method for RflomicsSE</span></span>
<span><span class="fu">plotKEGG</span><span class="op">(</span></span>
<span>  <span class="va">object</span>,</span>
<span>  <span class="va">contrastName</span>,</span>
<span>  pathway_id <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  species <span class="op">=</span> <span class="st">"ath"</span>,</span>
<span>  gene_idtype <span class="op">=</span> <span class="st">"kegg"</span>,</span>
<span>  from <span class="op">=</span> <span class="st">"DiffExp"</span>,</span>
<span>  <span class="va">...</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># S4 method for RflomicsSE</span></span>
<span><span class="fu">plotClusterProfiler</span><span class="op">(</span></span>
<span>  <span class="va">object</span>,</span>
<span>  <span class="va">contrastName</span>,</span>
<span>  <span class="va">database</span>,</span>
<span>  domain <span class="op">=</span> <span class="st">"no-domain"</span>,</span>
<span>  from <span class="op">=</span> <span class="st">"DiffExp"</span>,</span>
<span>  plotType <span class="op">=</span> <span class="st">"dotplot"</span>,</span>
<span>  showCategory <span class="op">=</span> <span class="fl">15</span>,</span>
<span>  searchExpr <span class="op">=</span> <span class="st">""</span>,</span>
<span>  nodeLabel <span class="op">=</span> <span class="st">"all"</span>,</span>
<span>  p.adj.cutoff <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  <span class="va">...</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># S4 method for RflomicsSE</span></span>
<span><span class="fu">plotEnrichComp</span><span class="op">(</span></span>
<span>  <span class="va">object</span>,</span>
<span>  from <span class="op">=</span> <span class="st">"DiffExp"</span>,</span>
<span>  database <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  domain <span class="op">=</span> <span class="st">"no-domain"</span>,</span>
<span>  matrixType <span class="op">=</span> <span class="st">"FC"</span>,</span>
<span>  nClust <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  <span class="va">...</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># S4 method for RflomicsSE</span></span>
<span><span class="fu">getEnrichRes</span><span class="op">(</span></span>
<span>  <span class="va">object</span>,</span>
<span>  contrastName <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  from <span class="op">=</span> <span class="st">"DiffExpEnrichAnal"</span>,</span>
<span>  database <span class="op">=</span> <span class="st">"GO"</span>,</span>
<span>  domain <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  <span class="va">...</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># S4 method for RflomicsMAE</span></span>
<span><span class="fu">getEnrichRes</span><span class="op">(</span></span>
<span>  <span class="va">object</span>,</span>
<span>  <span class="va">experiment</span>,</span>
<span>  contrastName <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  from <span class="op">=</span> <span class="st">"DiffExpEnrichAnal"</span>,</span>
<span>  database <span class="op">=</span> <span class="st">"GO"</span>,</span>
<span>  domain <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  <span class="va">...</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># S4 method for RflomicsSE</span></span>
<span><span class="fu">sumORA</span><span class="op">(</span></span>
<span>  <span class="va">object</span>,</span>
<span>  from <span class="op">=</span> <span class="st">"DiffExpEnrichAnal"</span>,</span>
<span>  database <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  contrastName <span class="op">=</span> <span class="cn">NULL</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># S4 method for RflomicsSE</span></span>
<span><span class="fu">getEnrichPvalue</span><span class="op">(</span><span class="va">object</span>, from <span class="op">=</span> <span class="st">"DiffExpEnrichAnal"</span>, database <span class="op">=</span> <span class="st">"GO"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># S4 method for RflomicsSE</span></span>
<span><span class="fu">getEnrichSettings</span><span class="op">(</span><span class="va">object</span>, from <span class="op">=</span> <span class="st">"DiffExpEnrichAnal"</span>, database <span class="op">=</span> <span class="st">"GO"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># S4 method for RflomicsMAE</span></span>
<span><span class="fu">getAnnotAnalysesSummary</span><span class="op">(</span></span>
<span>  <span class="va">object</span>,</span>
<span>  from <span class="op">=</span> <span class="st">"DiffExpEnrichAnal"</span>,</span>
<span>  matrixType <span class="op">=</span> <span class="st">"presence"</span>,</span>
<span>  <span class="va">...</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>object</dt>
<dd><p>An object of class <a href="RflomicsSE-class.html">RflomicsSE</a> or 
class <a href="RflomicsMAE-class.html">RflomicsMAE-class</a></p></dd>


<dt>nameList</dt>
<dd><p>name of contrasts (tags or names) from which to extract DE
genes if from is DiffExpAnal.</p></dd>


<dt>list_args</dt>
<dd><p>list of arguments to pass to the enrichment function.
These arguments must match the ones from the clusterprofiler package.
E.g: universe, keytype, pvalueCutoff, qvalueCutoff, etc.</p></dd>


<dt>from</dt>
<dd><p>indicates if the enrichment results are taken from differential 
analysis results (DiffExpEnrichAnal) or from the co-expression analysis 
results (CoExpEnrichAnal)</p></dd>


<dt>database</dt>
<dd><p>which database (GO, KEGG, custom...)</p></dd>


<dt>domain</dt>
<dd><p>the subonology or subdomain for the database (eg CC, MF or
BP for GO.)</p></dd>


<dt>col_term</dt>
<dd><p>for custom annotation, column name of the file containing
term names or id.</p></dd>


<dt>col_gene</dt>
<dd><p>for custom annotation, column name of the file containing
entities names.</p></dd>


<dt>col_name</dt>
<dd><p>(optional) for custom annotation, column name of the 
file containing term names if col_term is used for ids.</p></dd>


<dt>col_domain</dt>
<dd><p>for custom annotation, column name of the file containing
the domains.</p></dd>


<dt>annot</dt>
<dd><p>for custom annotation, path of the annotation file.</p></dd>


<dt>SE.name</dt>
<dd><p>SE.name the name of the dataset if the input object 
is a <a href="RflomicsMAE-class.html">RflomicsMAE-class</a></p></dd>


<dt>contrastName</dt>
<dd><p>the contrastName or clusterName to retrieve 
the results from. If NULL, all results are returned.</p></dd>


<dt>pathway_id</dt>
<dd><p>the KEGG id pathway to plot.</p></dd>


<dt>species</dt>
<dd><p>Kegg code (eg hsa or ath)</p></dd>


<dt>gene_idtype</dt>
<dd><p>idtype (kegg, uniprot,...)</p></dd>


<dt>...</dt>
<dd><p>more arguments for ComplexHeatmap::Heatmap.</p></dd>


<dt>plotType</dt>
<dd><p>type of plot. Define the function used inside.
One of dotplot, heatplot or cnetplot.</p></dd>


<dt>showCategory</dt>
<dd><p>max number of terms to show.</p></dd>


<dt>searchExpr</dt>
<dd><p>expression to search in the showCategory terms.</p></dd>


<dt>nodeLabel</dt>
<dd><p>same as in enrichplot::cnetplot function, defines</p></dd>


<dt>p.adj.cutoff</dt>
<dd><p>pvalueCutoff to define the enrichment threshold.</p></dd>


<dt>matrixType</dt>
<dd><p>Heatmap matrix to plot, one of GeneRatio, p.adjust 
or presence.</p></dd>


<dt>nClust</dt>
<dd><p>number of separate cluster to plot on the heatmap, based o
n the clustering.</p></dd>


<dt>experiment</dt>
<dd><p>if the object is a RflomicsMAE, then experiment is the
name of the RflomicsSE to look for.</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    

<p>A RflomicsMAE or a RflomicsS, depending on the class of object
parameter. The enrichment results are added to the metadata slot, either
in DiffExpEnrichAnal or CoExpEnrichAnal.</p>


<p>Only displays the KEGG pathway, it does not return any object.</p>


<p>a list of tables or a table</p>


<p>the pvalue cutoff used for the analysis.</p>


<p>a list with all settings</p>
    </div>
    <div id="accessors">
    <h2>Accessors</h2>
    
 
<p>A set of getters and setters generic functions to access and 
modify objects of the slot metadata of a <a href="RflomicsMAE-class.html">RflomicsMAE-class</a> object or 
a <a href="RflomicsMAE-class.html">RflomicsMAE-class</a> object.</p>
 
<ul><li><p>getEnrichRes: get a particular enrichment result.
   return enrichment results given in the form of lists of clusterprofiler
   results.</p></li>
</ul><ul><li><p>sumORA: Get summary tables from ORA analyses -
   once an enrichment has been conducted.</p></li>
</ul><ul><li><p>getEnrichPvalue:
   Get the pvalue threshold used in enrichment analysis.</p></li>
</ul><ul><li><p>getEnrichSettings:
   get the settings of an enrichment analysis.</p></li>
</ul><ul><li><p>getAnnotAnalysesSummary:
   return A list of heatmaps, one for each ontology/domain.</p></li>
</ul></div>
    <div id="plots">
    <h2>Plots</h2>
    
 
<p>A collection of functions for plotting results from omic analysis steps.</p>
 
<ul><li><p>plotKEGG: Plot the KEGG pathway using the pathview package. 
   It overlays the result of the differential analysis or the clusters 
   entity on the image.</p></li>
</ul><ul><li><p>plotClusterProfiler: Plot a dotplot, a cnetplot or an heatplot, using enrichplot
package. It is a wrapper method destined for the RflomicsSE class..</p></li>
</ul><ul><li><p>plotEnrichComp: plot an heatmap of all the enriched term found for a given
database and a given source (differential analysis or coexpression clusters).
Allow for the comparison of several enrichment results.</p></li>
</ul></div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="co"># load ecoseed data</span></span></span>
<span class="r-in"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://forgemia.inra.fr/flomics/rflomics/" class="external-link">RFLOMICS</a></span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">ecoseed</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># create rflomicsMAE object with ecoseed data</span></span></span>
<span class="r-in"><span><span class="va">MAE</span> <span class="op">&lt;-</span> <span class="fu"><a href="createRflomicsMAE.html">createRflomicsMAE</a></span><span class="op">(</span></span></span>
<span class="r-in"><span>    projectName <span class="op">=</span> <span class="st">"Tests"</span>,</span></span>
<span class="r-in"><span>    omicsData   <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">ecoseed</span><span class="op">$</span><span class="va">RNAtest</span>, <span class="va">ecoseed</span><span class="op">$</span><span class="va">metatest</span>, <span class="va">ecoseed</span><span class="op">$</span><span class="va">protetest</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>    omicsNames  <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"RNAtest"</span>, <span class="st">"metatest"</span>, <span class="st">"protetest"</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>    omicsTypes  <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"RNAseq"</span>,<span class="st">"metabolomics"</span>,<span class="st">"proteomics"</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>    ExpDesign   <span class="op">=</span> <span class="va">ecoseed</span><span class="op">$</span><span class="va">design</span>,</span></span>
<span class="r-in"><span>    factorRef   <span class="op">=</span> <span class="va">ecoseed</span><span class="op">$</span><span class="va">factorRef</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="co"># names(MAE) &lt;- c("RNAtest", "metatest", "protetest")</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Set the statistical model and contrasts to test</span></span></span>
<span class="r-in"><span><span class="va">formulae</span> <span class="op">&lt;-</span> <span class="fu"><a href="generateModelFormulae.html">generateModelFormulae</a></span><span class="op">(</span><span class="va">MAE</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">MAE</span> <span class="op">&lt;-</span> <span class="fu"><a href="generateModelFormulae.html">setModelFormula</a></span><span class="op">(</span><span class="va">MAE</span>, <span class="va">formulae</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">)</span>  </span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Get the contrasts List and choose the first 3 contrasts of type averaged</span></span></span>
<span class="r-in"><span><span class="va">contrastList</span> <span class="op">&lt;-</span> <span class="fu"><a href="generateExpressionContrast.html">generateExpressionContrast</a></span><span class="op">(</span><span class="va">MAE</span>, <span class="st">"averaged"</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">MAE</span> <span class="op">&lt;-</span> <span class="fu"><a href="generateExpressionContrast.html">setSelectedContrasts</a></span><span class="op">(</span><span class="va">MAE</span>, contrastList <span class="op">=</span> <span class="va">contrastList</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">2</span>, <span class="fl">3</span><span class="op">)</span>,<span class="op">]</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Run the data preprocessing and perform the differential analysis </span></span></span>
<span class="r-in"><span><span class="va">MAE</span> <span class="op">&lt;-</span> <span class="fu"><a href="runDataProcessing.html">runDataProcessing</a></span><span class="op">(</span><span class="va">MAE</span>, SE.name <span class="op">=</span> <span class="st">"protetest"</span>,  </span></span>
<span class="r-in"><span>                         transformMethod <span class="op">=</span> <span class="st">"log2"</span>,</span></span>
<span class="r-in"><span>                         normMethod <span class="op">=</span> <span class="st">"median"</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> [RFLOMICS] #    =&gt; select samples...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> [RFLOMICS] #    =&gt; transformation data...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> [RFLOMICS] #    =&gt; Run normalization...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> [RFLOMICS] #    =&gt; Compute PCA </span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">MAE</span> <span class="op">&lt;-</span> <span class="fu"><a href="runDiffAnalysis.html">runDiffAnalysis</a></span><span class="op">(</span><span class="va">MAE</span>, </span></span>
<span class="r-in"><span>                       SE.name <span class="op">=</span> <span class="st">"protetest"</span>, </span></span>
<span class="r-in"><span>                       method <span class="op">=</span> <span class="st">"limmalmFit"</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Run GO annotation (enrichGO)</span></span></span>
<span class="r-in"><span><span class="co"># Not run: need org.At.tair.db package</span></span></span>
<span class="r-in"><span><span class="co"># MAE &lt;- runAnnotationEnrichment(MAE, SE.name = "protetest",</span></span></span>
<span class="r-in"><span><span class="co">#                                list_args = list(OrgDb = "org.At.tair.db",</span></span></span>
<span class="r-in"><span><span class="co">#                                                 keyType = "TAIR",</span></span></span>
<span class="r-in"><span><span class="co">#                                                 pvalueCutoff = 0.05),</span></span></span>
<span class="r-in"><span><span class="co">#                                from = "DiffExp", database = "GO",</span></span></span>
<span class="r-in"><span><span class="co">#                                domain = "CC")</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Run KEGG annotation (enrichKEGG)</span></span></span>
<span class="r-in"><span><span class="co"># need internet connection</span></span></span>
<span class="r-in"><span><span class="va">MAE</span> <span class="op">&lt;-</span> <span class="fu">runAnnotationEnrichment</span><span class="op">(</span><span class="va">MAE</span>, SE.name <span class="op">=</span> <span class="st">"protetest"</span>,</span></span>
<span class="r-in"><span>                               list_args <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>organism <span class="op">=</span> <span class="st">"ath"</span>,</span></span>
<span class="r-in"><span>                                                keyType <span class="op">=</span> <span class="st">"kegg"</span>,</span></span>
<span class="r-in"><span>                                                pvalueCutoff <span class="op">=</span> <span class="fl">0.05</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>                               from <span class="op">=</span> <span class="st">"DiffExp"</span>, database <span class="op">=</span> <span class="st">"KEGG"</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> warning: </span>
<span class="r-in"><span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Search for the pvalue cutoff:</span></span></span>
<span class="r-in"><span><span class="fu">sumORA</span><span class="op">(</span><span class="va">MAE</span><span class="op">[[</span><span class="st">"protetest"</span><span class="op">]</span><span class="op">]</span>, from <span class="op">=</span> <span class="st">"DiffExp"</span>, database <span class="op">=</span> <span class="st">"KEGG"</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                                                                                            Contrast</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> (temperatureMedium - temperatureLow) in mean           (temperatureMedium - temperatureLow) in mean</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> (temperatureElevated - temperatureLow) in mean       (temperatureElevated - temperatureLow) in mean</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> (temperatureElevated - temperatureMedium) in mean (temperatureElevated - temperatureMedium) in mean</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                                                   no-domain</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> (temperatureMedium - temperatureLow) in mean              5</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> (temperatureElevated - temperatureLow) in mean           31</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> (temperatureElevated - temperatureMedium) in mean        23</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># need internet connection</span></span></span>
<span class="r-in"><span><span class="co"># pathview package</span></span></span>
<span class="r-in"><span><span class="co"># plotKEGG(MAE[["protetest"]], pathway_id = "ath00710", species = "ath",</span></span></span>
<span class="r-in"><span><span class="co">#          contrastName = "cluster.4", from = "Coexp")</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># From differential analysis proteins lists:</span></span></span>
<span class="r-in"><span><span class="co"># plotClusterProfiler(MAE[["protetest"]],</span></span></span>
<span class="r-in"><span><span class="co">#                     contrastName = "(temperatureElevated - temperatureMedium) in mean",</span></span></span>
<span class="r-in"><span><span class="co">#                     database = "KEGG", from = "DiffExp",</span></span></span>
<span class="r-in"><span><span class="co">#                     plotType = "heatplot", p.adj.cutoff = 0.05,</span></span></span>
<span class="r-in"><span><span class="co">#                     domain = "no-domain")</span></span></span>
<span class="r-in"><span><span class="co"># </span></span></span>
<span class="r-in"><span><span class="co"># plotEnrichComp(MAE[["protetest"]], from = "DiffExp",</span></span></span>
<span class="r-in"><span><span class="co">#                database = "KEGG", matrixType = "FC")</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Get all results from KEGG on differential expression lists:</span></span></span>
<span class="r-in"><span><span class="va">results</span> <span class="op">&lt;-</span> <span class="fu">getEnrichRes</span><span class="op">(</span><span class="va">MAE</span><span class="op">[[</span><span class="st">"protetest"</span><span class="op">]</span><span class="op">]</span>,</span></span>
<span class="r-in"><span>                        from <span class="op">=</span> <span class="st">"diffexp"</span>, database <span class="op">=</span> <span class="st">"KEGG"</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Search for the pvalue cutoff:</span></span></span>
<span class="r-in"><span><span class="va">usedPvalue</span> <span class="op">&lt;-</span> </span></span>
<span class="r-in"><span>    <span class="fu">getEnrichPvalue</span><span class="op">(</span><span class="va">MAE</span><span class="op">[[</span><span class="st">"protetest"</span><span class="op">]</span><span class="op">]</span>, from <span class="op">=</span> <span class="st">"diffexp"</span>, database <span class="op">=</span> <span class="st">"KEGG"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">settings</span> <span class="op">&lt;-</span> </span></span>
<span class="r-in"><span>    <span class="fu">getEnrichSettings</span><span class="op">(</span><span class="va">MAE</span><span class="op">[[</span><span class="st">"protetest"</span><span class="op">]</span><span class="op">]</span>, from <span class="op">=</span> <span class="st">"diffexp"</span>, database <span class="op">=</span> <span class="st">"KEGG"</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Nadia Bessoltane, Delphine Charif, Audrey Hulot, Christine Paysant-Leroux, Gwendal Cueff.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.9.</p>
</div>

      </footer></div>

  


  

  </body></html>

