<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>runCoExpression — runCoExpression • RFLOMICS</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="runCoExpression — runCoExpression"><meta property="og:description" content="This method performs a co-expression/co-abundance analysis of 
omic-data."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">RFLOMICS</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.0.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../articles/RFLOMICS.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Manuals
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/RFLOMICS-cmd-line.html">RFLOMICS vignettes cmd line</a>
    </li>
    <li>
      <a href="../articles/RFLOMICS-input-data.html">RFLOMICS input data</a>
    </li>
  </ul></li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://forgemia.inra.fr/flomics/rflomics" class="external-link">
    <span class="fa fa-gitlab"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>runCoExpression</h1>
    
    <div class="hidden name"><code>runCoExpression.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>This method performs a co-expression/co-abundance analysis of 
omic-data.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="co"># S4 method for RflomicsSE</span></span>
<span><span class="fu">runCoExpression</span><span class="op">(</span></span>
<span>  <span class="va">object</span>,</span>
<span>  K <span class="op">=</span> <span class="fl">2</span><span class="op">:</span><span class="fl">20</span>,</span>
<span>  replicates <span class="op">=</span> <span class="fl">5</span>,</span>
<span>  contrastNames <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  merge <span class="op">=</span> <span class="st">"union"</span>,</span>
<span>  model <span class="op">=</span> <span class="st">"Normal"</span>,</span>
<span>  GaussianModel <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  transformation <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  normFactors <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  clustermq <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  meanFilterCutoff <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  scale <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  silent <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  cmd <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># S4 method for RflomicsMAE</span></span>
<span><span class="fu">runCoExpression</span><span class="op">(</span></span>
<span>  <span class="va">object</span>,</span>
<span>  <span class="va">SE.name</span>,</span>
<span>  K <span class="op">=</span> <span class="fl">2</span><span class="op">:</span><span class="fl">20</span>,</span>
<span>  replicates <span class="op">=</span> <span class="fl">5</span>,</span>
<span>  <span class="va">contrastNames</span>,</span>
<span>  merge <span class="op">=</span> <span class="st">"union"</span>,</span>
<span>  model <span class="op">=</span> <span class="st">"Normal"</span>,</span>
<span>  GaussianModel <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  transformation <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  normFactors <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  clustermq <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  meanFilterCutoff <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  scale <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  silent <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  cmd <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># S4 method for RflomicsMAE</span></span>
<span><span class="fu">getCoExpAnalysesSummary</span><span class="op">(</span><span class="va">object</span>, omicNames <span class="op">=</span> <span class="cn">NULL</span>, <span class="va">...</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># S4 method for RflomicsSE</span></span>
<span><span class="fu">plotCoExpression</span><span class="op">(</span><span class="va">object</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># S4 method for RflomicsMAE</span></span>
<span><span class="fu">plotCoExpression</span><span class="op">(</span><span class="va">object</span>, <span class="va">SE.name</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># S4 method for RflomicsSE</span></span>
<span><span class="fu">plotCoExpressionProfile</span><span class="op">(</span></span>
<span>  <span class="va">object</span>,</span>
<span>  cluster <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  condition <span class="op">=</span> <span class="st">"groups"</span>,</span>
<span>  features <span class="op">=</span> <span class="cn">NULL</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># S4 method for RflomicsMAE</span></span>
<span><span class="fu">plotCoExpressionProfile</span><span class="op">(</span></span>
<span>  <span class="va">object</span>,</span>
<span>  <span class="va">SE.name</span>,</span>
<span>  cluster <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  condition <span class="op">=</span> <span class="st">"groups"</span>,</span>
<span>  features <span class="op">=</span> <span class="cn">NULL</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># S4 method for RflomicsSE</span></span>
<span><span class="fu">plotCoseqContrasts</span><span class="op">(</span><span class="va">object</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># S4 method for RflomicsMAE</span></span>
<span><span class="fu">plotCoseqContrasts</span><span class="op">(</span><span class="va">object</span>, <span class="va">SE.name</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># S4 method for RflomicsSE</span></span>
<span><span class="fu">getCoexpSettings</span><span class="op">(</span><span class="va">object</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># S4 method for RflomicsMAE</span></span>
<span><span class="fu">getCoexpSettings</span><span class="op">(</span><span class="va">object</span>, <span class="va">SE.name</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># S4 method for RflomicsSE</span></span>
<span><span class="fu">getClusterEntities</span><span class="op">(</span><span class="va">object</span>, <span class="va">clusterName</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># S4 method for RflomicsMAE</span></span>
<span><span class="fu">getClusterEntities</span><span class="op">(</span><span class="va">object</span>, <span class="va">SE.name</span>, <span class="va">clusterName</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># S4 method for RflomicsSE</span></span>
<span><span class="fu">getCoseqClusters</span><span class="op">(</span><span class="va">object</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># S4 method for RflomicsMAE</span></span>
<span><span class="fu">getCoseqClusters</span><span class="op">(</span><span class="va">object</span>, <span class="va">SE.name</span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>object</dt>
<dd><p>An object of class <a href="RflomicsSE-class.html">RflomicsSE</a> or 
class <a href="RflomicsMAE-class.html">RflomicsMAE-class</a></p></dd>


<dt>K</dt>
<dd><p>Number of clusters (a single value or a vector of values)</p></dd>


<dt>replicates</dt>
<dd><p>The number of iteration for each K.</p></dd>


<dt>contrastNames</dt>
<dd><p>names of the contrasts from which the DE entities 
have to be taken. Can be NULL, in that case every contrasts from the differential 
analysis are taken into consideration.</p></dd>


<dt>merge</dt>
<dd><p><code>"union"</code> or <code>"intersection"</code></p></dd>


<dt>model</dt>
<dd><p>Type of mixture model to use <code>"Poisson"</code> or 
<code>"normal"</code>. By default, it is the normal.</p></dd>


<dt>GaussianModel</dt>
<dd><p>Type of <code>GaussianModel</code> to be used for the 
Normal mixture model only. This parameters
is set to <code>"Gaussian_pk_Lk_Ck"</code> by default and doesn't have to 
be changed except if an error message proposed
to try another model like <code>"Gaussian_pk_Lk_Bk"</code>.</p></dd>


<dt>transformation</dt>
<dd><p>The transformation type to be used. By default, 
it is the "arcsin" one.</p></dd>


<dt>normFactors</dt>
<dd><p>The type of estimator to be used to normalize for 
differences in library size.
By default, it is the "TMM" one.</p></dd>


<dt>clustermq</dt>
<dd><p>boolean. Does the computation need to be executed 
on a distant server. A configuration file and a network connection are 
needed to use this option.</p></dd>


<dt>meanFilterCutoff</dt>
<dd><p>a cutoff to filter a gene with a mean expression lower than this value.
(only for RNAseq data, set to NULL for others).</p></dd>


<dt>scale</dt>
<dd><p>Boolean. If TRUE scale all variables tounit variance.</p></dd>


<dt>silent</dt>
<dd><p>if TRUE, coseq run silently (without any console print or 
message)</p></dd>


<dt>cmd</dt>
<dd><p>if TRUE, print steps of the analysis. Used inside the coseq 
module in the shiny interface.</p></dd>


<dt>SE.name</dt>
<dd><p>SE.name the name of the dataset if the input object 
is a <a href="RflomicsMAE-class.html">RflomicsMAE-class</a></p></dd>


<dt>omicNames</dt>
<dd><p>the name of the experiment to summarize.</p></dd>


<dt>...</dt>
<dd><p>Not in use at the moment</p></dd>


<dt>cluster</dt>
<dd><p>cluster number</p></dd>


<dt>condition</dt>
<dd><p>Default is group.</p></dd>


<dt>features</dt>
<dd><p>Default is NULL.</p></dd>


<dt>clusterName</dt>
<dd><p>name of the cluster</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    

<p>An S4 object of class <a href="RflomicsSE-class.html">RflomicsSE</a></p>


<p>All the results are stored as a named list <code>CoExpAnal</code></p>


<p>in the metadata slot of a given <code>RflomicsSE</code> object. Objects are:
The runCoExpression method return several results, for coseq</p>
 

<p>method, objects are:</p><ul><li><p><code>setting:</code>  co-expression analysis settings. See <code>getCoexpSetting</code></p></li>
<li><p><code>results:</code>  boolean indicating if the co-expression analysis succeed</p></li>
<li><p><code>coseqResults:</code> the raw results of <code>coseq</code></p></li>
<li><p><code>clusters:</code>  a List of clusters</p></li>
<li><p><code>cluster.nb:</code>  The number of cluster</p></li>
<li><p><code>plots:</code>  The plots of <code>coseq</code> results</p></li>
<li><p><code>stats:</code>  A tibble summarising failed jobs: reason, propoif any</p></li>
</ul></div>
    <div id="details">
    <h2>Details</h2>
    <p>For now, only the coseq function of the coseq package is used.
For RNAseq data, parameters used are those recommended in DiCoExpress 
workflow (see the reference). This parameters are: <code>model="normal"</code>, 
<code>transformation="arcsin"</code>, <code>GaussianModel="Gaussian_pk_Lk_Ck"</code>,
<code>normFactors="TMM"</code>, <code>meanFilterCutoff = 50</code>
For proteomic or metabolomic, data are scaled by protein or metabolite 
to group them by expression profiles rather than by expression intensity.
After data scaling, recommended parameters (from <code>coseq</code> developers) 
for co-expression analysis are:
<code>model="normal"</code>, <code>transformation="none"</code>, 
<code>GaussianModel="Gaussian_pk_Lk_Ck"</code>,
<code>normFactors="none"</code>,  <code>meanFilterCutoff = NULL</code>.</p>
    </div>
    <div id="accessors">
    <h2>Accessors</h2>
    
 
<p>A set of getters and setters generic functions to access and 
modify objects of the slot metadata of a <a href="RflomicsMAE-class.html">RflomicsMAE-class</a> object or 
a <a href="RflomicsMAE-class.html">RflomicsMAE-class</a> object.</p>
 
<ul><li><p>getCoexpSettings: Access to the co-expression analysis settings 
   of a given omic dataset</p></li>
</ul><ul><li><p>getClusterEntities: get members of a cluster. 
 return The list of entities inside this cluster.</p></li>
</ul><ul><li><p>getCoseqClusters: get cluster. Return all clusters</p></li>
</ul></div>
    <div id="plots">
    <h2>Plots</h2>
    
 
<p>A collection of functions for plotting results from omic analysis steps.</p>
 
<ul><li><p>getCoExpAnalysesSummary: ...</p></li>
</ul><ul><li><p>plotCoExpression: 
   list plot of ICL, logLike and coseq object with min ICL</p></li>
</ul><ul><li><p>plotCoExpressionProfile: 
   ...</p></li>
</ul><ul><li><p>plotCoseqContrasts: This function describes the composition of 
   clusters according to the contrast to which the gene belongs</p></li>
</ul></div>
    <div id="references">
    <h2>References</h2>
    <p>Lambert, I., Paysant-Le Roux, C., Colella, S. et al. DiCoExpress: 
a tool to process multifactorial RNAseq experiments from quality controls 
to co-expression analysis through differential analysis based on contrasts 
inside GLM models. Plant Methods 16, 68 (2020).</p>
    </div>
    <div id="see-also">
    <h2>See also</h2>
    <div class="dont-index"><p><code><a href="https://rdrr.io/pkg/coseq/man/coseq.html" class="external-link">coseq</a></code></p>
<p><code><a href="createRflomicsMAE.html">createRflomicsMAE</a></code></p>
<p><code><a href="generateModelFormulae.html">generateModelFormulae</a></code></p>
<p><code><a href="generateExpressionContrast.html">generateExpressionContrast</a></code></p>
<p><code><a href="runDataProcessing.html">runDataProcessing</a></code></p>
<p><code><a href="runDiffAnalysis.html">runDiffAnalysis</a></code></p></div>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="co"># load ecoseed data</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">ecoseed</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># create rflomicsMAE object with ecoseed data</span></span></span>
<span class="r-in"><span><span class="va">MAE</span> <span class="op">&lt;-</span> <span class="fu">RFLOMICS</span><span class="fu">::</span><span class="fu"><a href="createRflomicsMAE.html">createRflomicsMAE</a></span><span class="op">(</span></span></span>
<span class="r-in"><span>  projectName <span class="op">=</span> <span class="st">"Tests"</span>,</span></span>
<span class="r-in"><span>  omicsData   <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">ecoseed</span><span class="op">$</span><span class="va">RNAtest</span>, <span class="va">ecoseed</span><span class="op">$</span><span class="va">metatest</span>, <span class="va">ecoseed</span><span class="op">$</span><span class="va">protetest</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>  omicsNames  <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"RNAtest"</span>, <span class="st">"metatest"</span>, <span class="st">"protetest"</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>  omicsTypes  <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"RNAseq"</span>,<span class="st">"metabolomics"</span>,<span class="st">"proteomics"</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>  ExpDesign   <span class="op">=</span> <span class="va">ecoseed</span><span class="op">$</span><span class="va">design</span>,</span></span>
<span class="r-in"><span>  factorRef   <span class="op">=</span> <span class="va">ecoseed</span><span class="op">$</span><span class="va">factorRef</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Set the statistical model and contrasts to test</span></span></span>
<span class="r-in"><span><span class="va">formulae</span> <span class="op">&lt;-</span> <span class="fu"><a href="generateModelFormulae.html">generateModelFormulae</a></span><span class="op">(</span><span class="va">MAE</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">MAE</span> <span class="op">&lt;-</span> <span class="fu"><a href="generateModelFormulae.html">setModelFormula</a></span><span class="op">(</span><span class="va">MAE</span>, <span class="va">formulae</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">)</span>  </span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Get the contrasts List and choose the first 3 contrasts of type averaged</span></span></span>
<span class="r-in"><span><span class="va">contrastList</span> <span class="op">&lt;-</span> <span class="fu"><a href="generateExpressionContrast.html">generateExpressionContrast</a></span><span class="op">(</span><span class="va">MAE</span>, <span class="st">"averaged"</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">MAE</span> <span class="op">&lt;-</span> <span class="fu"><a href="generateExpressionContrast.html">setSelectedContrasts</a></span><span class="op">(</span><span class="va">MAE</span>, contrastList <span class="op">=</span> <span class="va">contrastList</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">2</span>, <span class="fl">3</span><span class="op">)</span>,<span class="op">]</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Run the data preprocessing and perform the differential analysis </span></span></span>
<span class="r-in"><span><span class="va">MAE</span> <span class="op">&lt;-</span> <span class="fu"><a href="runDataProcessing.html">runDataProcessing</a></span><span class="op">(</span><span class="va">MAE</span>, SE.name <span class="op">=</span> <span class="st">"protetest"</span>,  </span></span>
<span class="r-in"><span>                         transformMethod <span class="op">=</span> <span class="st">"log2"</span>,</span></span>
<span class="r-in"><span>                         normMethod <span class="op">=</span> <span class="st">"median"</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> [RFLOMICS] #    =&gt; select samples...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> [RFLOMICS] #    =&gt; transformation data...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> [RFLOMICS] #    =&gt; Run normalization...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> [RFLOMICS] #    =&gt; Compute PCA </span>
<span class="r-in"><span><span class="va">MAE</span> <span class="op">&lt;-</span> <span class="fu"><a href="runDiffAnalysis.html">runDiffAnalysis</a></span><span class="op">(</span><span class="va">MAE</span>, SE.name <span class="op">=</span> <span class="st">"protetest"</span><span class="op">)</span></span></span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>DiffAnalyseMethod was missing. Detected omic type is proteomics using limmalmFit for differential analysis.</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Run co-expression analysis</span></span></span>
<span class="r-in"><span><span class="va">MAE</span> <span class="op">&lt;-</span> <span class="fu">runCoExpression</span><span class="op">(</span><span class="va">MAE</span>, SE.name <span class="op">=</span> <span class="st">"protetest"</span>, </span></span>
<span class="r-in"><span>                       K <span class="op">=</span> <span class="fl">2</span><span class="op">:</span><span class="fl">5</span>, replicates <span class="op">=</span> <span class="fl">5</span>, </span></span>
<span class="r-in"><span>                       merge <span class="op">=</span> <span class="st">"union"</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># get parametres used to run co-expression analysis</span></span></span>
<span class="r-in"><span><span class="va">coExp.set.list</span> <span class="op">&lt;-</span> <span class="fu">getCoexpSettings</span><span class="op">(</span><span class="va">MAE</span><span class="op">[[</span><span class="st">"protetest"</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">coExp.set.list</span><span class="op">$</span><span class="va">method</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "coseq"</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># get results</span></span></span>
<span class="r-in"><span><span class="va">clusters</span> <span class="op">&lt;-</span> <span class="fu">getCoseqClusters</span><span class="op">(</span><span class="va">MAE</span><span class="op">[[</span><span class="st">"protetest"</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">proteins_c1</span> <span class="op">&lt;-</span> <span class="fu">getClusterEntities</span><span class="op">(</span><span class="va">MAE</span><span class="op">[[</span><span class="st">"protetest"</span><span class="op">]</span><span class="op">]</span>, </span></span>
<span class="r-in"><span>                                      clusterName <span class="op">=</span> <span class="st">"cluster_1"</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># plots</span></span></span>
<span class="r-in"><span><span class="co">#plotCoExpression(MAE[["protetest"]])</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">#plotCoExpressionProfile(MAE[["protetest"]], cluster = 2)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">#plotCoseqContrasts(MAE[["protetest"]])</span></span></span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Nadia Bessoltane, Delphine Charif, Audrey Hulot, Christine Paysant-Leroux, Gwendal Cueff.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.9.</p>
</div>

      </footer></div>

  


  

  </body></html>

