<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Run Differential Expression Analysis and process results — runDiffAnalysis • RFLOMICS</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Run Differential Expression Analysis and process results — runDiffAnalysis"><meta property="og:description" content="This is an interface method which run a differential 
analysis on omics datasets stored in an object of class 
RflomicsSE or RflomicsMAE-class. According to the type 
of omics and to a list of contrasts, a differential analysis  
is performef for each contrasts.
Two methods are available according to the type of object:
For RNAseq data:  the glmFit function/model of the 
edgeR package is applied.
For proteomic and metabolomic data:  the lmFit 
function/model of the limma package is applied.



generateContrastMatrix: 
 Defines contrast matrix or contrast list with contrast 
 name and contrast coefficients



filterDiffAnalysis: The filterDiffAnalysis method allows 
   filtering the results of the differential analysis based on a 
   new cutoff for p-value and fold change.



setValidContrasts: Set the valid contrasts stored in metadata slot.

"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">RFLOMICS</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.0.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../articles/RFLOMICS.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Manuals
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/RFLOMICS-cmd-line.html">RFLOMICS vignettes cmd line</a>
    </li>
    <li>
      <a href="../articles/RFLOMICS-input-data.html">RFLOMICS input data</a>
    </li>
  </ul></li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://forgemia.inra.fr/flomics/rflomics" class="external-link">
    <span class="fa fa-gitlab"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Run Differential Expression Analysis and process results</h1>
    
    <div class="hidden name"><code>runDiffAnalysis.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>This is an interface method which run a differential 
analysis on omics datasets stored in an object of class 
<a href="RflomicsSE-class.html">RflomicsSE</a> or <a href="RflomicsMAE-class.html">RflomicsMAE-class</a>. According to the type 
of omics and to a list of contrasts, a differential analysis  
is performef for each contrasts.
Two methods are available according to the type of object:</p><ul><li><p>For RNAseq data:  the <code>glmFit</code> function/model of the 
<code>edgeR</code> package is applied.</p></li>
<li><p>For proteomic and metabolomic data:  the <code>lmFit</code> 
function/model of the <code>limma</code> package is applied.</p></li>
</ul><ul><li><p>generateContrastMatrix: 
 Defines contrast matrix or contrast list with contrast 
 name and contrast coefficients</p></li>
</ul><ul><li><p>filterDiffAnalysis: The filterDiffAnalysis method allows 
   filtering the results of the differential analysis based on a 
   new cutoff for p-value and fold change.</p></li>
</ul><ul><li><p>setValidContrasts: Set the valid contrasts stored in <code>metadata</code> slot.</p></li>
</ul></div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="co"># S4 method for RflomicsSE</span></span>
<span><span class="fu">runDiffAnalysis</span><span class="op">(</span></span>
<span>  <span class="va">object</span>,</span>
<span>  contrastList <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  method <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  p.adj.method <span class="op">=</span> <span class="st">"BH"</span>,</span>
<span>  p.adj.cutoff <span class="op">=</span> <span class="fl">0.05</span>,</span>
<span>  logFC.cutoff <span class="op">=</span> <span class="fl">0</span>,</span>
<span>  clustermq <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  cmd <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  <span class="va">...</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># S4 method for RflomicsMAE</span></span>
<span><span class="fu">runDiffAnalysis</span><span class="op">(</span></span>
<span>  <span class="va">object</span>,</span>
<span>  <span class="va">SE.name</span>,</span>
<span>  contrastList <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  method <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  p.adj.method <span class="op">=</span> <span class="st">"BH"</span>,</span>
<span>  p.adj.cutoff <span class="op">=</span> <span class="fl">0.05</span>,</span>
<span>  logFC.cutoff <span class="op">=</span> <span class="fl">0</span>,</span>
<span>  clustermq <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  cmd <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  <span class="va">...</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># S4 method for RflomicsSE</span></span>
<span><span class="fu">generateContrastMatrix</span><span class="op">(</span><span class="va">object</span>, contrastList <span class="op">=</span> <span class="cn">NULL</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># S4 method for RflomicsMAE</span></span>
<span><span class="fu">generateContrastMatrix</span><span class="op">(</span><span class="va">object</span>, <span class="va">SE.name</span>, contrastList <span class="op">=</span> <span class="cn">NULL</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># S4 method for RflomicsSE</span></span>
<span><span class="fu">filterDiffAnalysis</span><span class="op">(</span><span class="va">object</span>, p.adj.cutoff <span class="op">=</span> <span class="fl">0.05</span>, logFC.cutoff <span class="op">=</span> <span class="fl">0</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># S4 method for RflomicsMAE</span></span>
<span><span class="fu">filterDiffAnalysis</span><span class="op">(</span><span class="va">object</span>, <span class="va">SE.name</span>, p.adj.cutoff <span class="op">=</span> <span class="cn">NULL</span>, logFC.cutoff <span class="op">=</span> <span class="cn">NULL</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># S4 method for RflomicsSE</span></span>
<span><span class="fu">setValidContrasts</span><span class="op">(</span><span class="va">object</span>, contrastList <span class="op">=</span> <span class="cn">NULL</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># S4 method for RflomicsMAE</span></span>
<span><span class="fu">setValidContrasts</span><span class="op">(</span><span class="va">object</span>, omicName <span class="op">=</span> <span class="cn">NULL</span>, contrastList <span class="op">=</span> <span class="cn">NULL</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># S4 method for RflomicsSE</span></span>
<span><span class="fu">plotDiffAnalysis</span><span class="op">(</span></span>
<span>  <span class="va">object</span>,</span>
<span>  <span class="va">contrastName</span>,</span>
<span>  typeofplots <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"MA.plot"</span>, <span class="st">"volcano"</span>, <span class="st">"histogram"</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># S4 method for RflomicsMAE</span></span>
<span><span class="fu">plotDiffAnalysis</span><span class="op">(</span></span>
<span>  <span class="va">object</span>,</span>
<span>  <span class="va">SE.name</span>,</span>
<span>  <span class="va">contrastName</span>,</span>
<span>  typeofplots <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"MA.plot"</span>, <span class="st">"volcano"</span>, <span class="st">"histogram"</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># S4 method for RflomicsSE</span></span>
<span><span class="fu">plotHeatmapDesign</span><span class="op">(</span></span>
<span>  <span class="va">object</span>,</span>
<span>  <span class="va">contrastName</span>,</span>
<span>  splitFactor <span class="op">=</span> <span class="st">"none"</span>,</span>
<span>  title <span class="op">=</span> <span class="st">""</span>,</span>
<span>  annotNames <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  modalities <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  drawArgs <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>  heatmapArgs <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># S4 method for RflomicsMAE</span></span>
<span><span class="fu">plotHeatmapDesign</span><span class="op">(</span></span>
<span>  <span class="va">object</span>,</span>
<span>  <span class="va">SE.name</span>,</span>
<span>  <span class="va">contrastName</span>,</span>
<span>  splitFactor <span class="op">=</span> <span class="st">"none"</span>,</span>
<span>  title <span class="op">=</span> <span class="st">""</span>,</span>
<span>  annotNames <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  modalities <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  drawArgs <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>  heatmapArgs <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># S4 method for RflomicsSE</span></span>
<span><span class="fu">plotBoxplotDE</span><span class="op">(</span><span class="va">object</span>, features <span class="op">=</span> <span class="cn">NULL</span>, groupColor <span class="op">=</span> <span class="st">"groups"</span>, raw <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># S4 method for RflomicsMAE</span></span>
<span><span class="fu">plotBoxplotDE</span><span class="op">(</span></span>
<span>  <span class="va">object</span>,</span>
<span>  <span class="va">SE.name</span>,</span>
<span>  features <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  groupColor <span class="op">=</span> <span class="st">"groups"</span>,</span>
<span>  raw <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># S4 method for RflomicsSE</span></span>
<span><span class="fu">getDEMatrix</span><span class="op">(</span><span class="va">object</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># S4 method for RflomicsMAE</span></span>
<span><span class="fu">getDEMatrix</span><span class="op">(</span><span class="va">object</span>, <span class="va">SE.name</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># S4 method for RflomicsSE</span></span>
<span><span class="fu">getDEList</span><span class="op">(</span><span class="va">object</span>, contrasts <span class="op">=</span> <span class="cn">NULL</span>, operation <span class="op">=</span> <span class="st">"union"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># S4 method for RflomicsMAE</span></span>
<span><span class="fu">getDEList</span><span class="op">(</span><span class="va">object</span>, <span class="va">SE.name</span>, contrasts <span class="op">=</span> <span class="cn">NULL</span>, operation <span class="op">=</span> <span class="st">"union"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># S4 method for RflomicsSE</span></span>
<span><span class="fu">getDiffSettings</span><span class="op">(</span><span class="va">object</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># S4 method for RflomicsMAE</span></span>
<span><span class="fu">getDiffSettings</span><span class="op">(</span><span class="va">object</span>, <span class="va">SE.name</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># S4 method for RflomicsSE</span></span>
<span><span class="fu">getValidContrasts</span><span class="op">(</span><span class="va">object</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># S4 method for RflomicsMAE</span></span>
<span><span class="fu">getValidContrasts</span><span class="op">(</span><span class="va">object</span>, <span class="va">omicName</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># S4 method for RflomicsSE</span></span>
<span><span class="fu">sumDiffExp</span><span class="op">(</span><span class="va">object</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># S4 method for RflomicsMAE</span></span>
<span><span class="fu">sumDiffExp</span><span class="op">(</span><span class="va">object</span>, SE.name <span class="op">=</span> <span class="cn">NULL</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># S4 method for RflomicsMAE</span></span>
<span><span class="fu">getDiffAnalysesSummary</span><span class="op">(</span></span>
<span>  <span class="va">object</span>,</span>
<span>  plot <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  ylabelLength <span class="op">=</span> <span class="fl">30</span>,</span>
<span>  nbMaxLabel <span class="op">=</span> <span class="fl">20</span>,</span>
<span>  interface <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>object</dt>
<dd><p>An object of class <a href="RflomicsSE-class.html">RflomicsSE</a> or 
class <a href="RflomicsMAE-class.html">RflomicsMAE-class</a></p></dd>


<dt>contrastList</dt>
<dd><p>A data.frame of contrast</p></dd>


<dt>method</dt>
<dd><p>A character vector giving the name of the differential 
analysis method to run. Either "edgeRglmfit" or "limmalmFit".</p></dd>


<dt>p.adj.method</dt>
<dd><p>The method choosen to adjust pvalue. Takes the same 
values as the ones of adj.p.adjust method.</p></dd>


<dt>p.adj.cutoff</dt>
<dd><p>adjusted pvalue cutoff. Default is the parameter from 
the differential analysis.</p></dd>


<dt>logFC.cutoff</dt>
<dd><p>cutoff for absolute value of log2FC. Default is the 
parameter from the differential analysis.</p></dd>


<dt>clustermq</dt>
<dd><p>A boolean indicating whether the constrasts have to be 
computed in local or in a distant machine</p></dd>


<dt>cmd</dt>
<dd><p>boolean. Used in the interface. If TRUE, print cmd for the user.</p></dd>


<dt>...</dt>
<dd><p>Additional arguments.</p></dd>


<dt>SE.name</dt>
<dd><p>SE.name the name of the dataset if the input object 
is a <a href="RflomicsMAE-class.html">RflomicsMAE-class</a></p></dd>


<dt>omicName</dt>
<dd><p>a dataset name</p></dd>


<dt>contrastName</dt>
<dd><p>The contrastName for which the MAplot has to be drawn</p></dd>


<dt>typeofplots</dt>
<dd><p>The plots you want to return. Default is all possible 
plots: MA plot, Volcano plot and non adjusted pvalues histogram.</p></dd>


<dt>splitFactor</dt>
<dd><p>characters. Default to none. Name of a feature in the 
design matrix, splits the samples on the heatmap according to its modalities.</p></dd>


<dt>title</dt>
<dd><p>characters. Title of the heatmap.</p></dd>


<dt>annotNames</dt>
<dd><p>vector. Names of the annotations to keep in the Heatmap. 
Default takes all available information.</p></dd>


<dt>modalities</dt>
<dd><p>named list of vectors of modalities to subset and print 
on the heatmap.</p></dd>


<dt>drawArgs, heatmapArgs</dt>
<dd><p>named lists. Any additional parameter passed 
to ComplexHeatmap::Heatmap or ComplexHeatmap::draw</p></dd>


<dt>features</dt>
<dd><p>variable name (gene/protein/metabolite name)</p></dd>


<dt>groupColor</dt>
<dd><p>default to groups, indicate a variable in the design to
color the boxplots accordingly.</p></dd>


<dt>raw</dt>
<dd><p>Boolean. Plot the raw data or the transformed ones (TRUE)</p></dd>


<dt>contrasts</dt>
<dd><p>Vector of characters, expect to be contrast names. 
Default is null, the operation (union) is performed
on every contrasts found.</p></dd>


<dt>operation</dt>
<dd><p>character. Either union or intersection.
Defines the operation to perform on the DE lists from the contrasts.</p></dd>


<dt>plot</dt>
<dd><p>FALSE or TRUE</p></dd>


<dt>ylabelLength</dt>
<dd><p>max length of the labels (characters)</p></dd>


<dt>nbMaxLabel</dt>
<dd><p>number of labels to print</p></dd>


<dt>interface</dt>
<dd><p>boolean. Is this plot for the interace or commandline?</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    

<p>A <a href="RflomicsSE-class.html">RflomicsSE</a> or a <a href="RflomicsMAE-class.html">RflomicsMAE-class</a> object.
All the results are stored as a named list <code>DiffExpAnal</code></p>
 

<p>in the metadata slot of a given <a href="RflomicsSE-class.html">RflomicsSE</a> object.
Objects are:</p><ul><li><p>stats: data.frame giving a summary of the differential 
 statistical analysis results by contrast:
 number of DE features, number of up and down regulated features</p></li>
<li><p>setting: Parameters used for the differential analysis</p></li>
<li><p>method: The method used for the differential analysis</p></li>
<li><p>p.adj.method: The applied p-value correction method</p></li>
<li><p>p.adj.cutoff: The cut-off applied for the adjusted p-value</p></li>
<li><p>logFC.cutoff: The absolute log FC cut-off</p></li>
<li><p>RawDEFres: a list giving for each contrast the raw results of 
 the differential analysis method</p></li>
<li><p>DEF: a list giving for each contrast a data.frame of non filtered 
 differential expressed features with their statistics</p></li>
<li><p>TopDEF: a list giving for each contrast a data.frame of 
 differential expressed features ordered and filtered by p.adj.cutoff 
 with their statistics</p></li>
<li><p>mergeDEF: a data frame of 0 and 1 indicating for each features in row, 
 if it is DE in a given contrasts in column</p></li>
<li><p>contrasts: a data.table of the contrasts used for the differential 
 analysis</p></li>
</ul><p>a data.frame with differential analyses summary</p>
    </div>
    <div id="details">
    <h2>Details</h2>
    <p>Functions and parameters used for RNAseq are those recommended in DiCoExpress 
workflow (see the paper in reference).
Functions and parameters used for proteomic and metabolomic data are those 
recommended in the (Efstathiou *et al.*, 2017)</p>
    </div>
    <div id="accessors">
    <h2>Accessors</h2>
    



<ul><li><p>getDEMatrix: return a matrix of experimental design.</p></li>
</ul><ul><li><p>getDEList: return a vector of union or intersection of differential
   expressed features from list of contrasts.</p></li>
</ul><ul><li><p>getDiffSettings: retrun a list of differential expression analysis 
   settings of a given omic dataset</p></li>
</ul><ul><li><p>getValidContrasts: return a data.frame of valided contrasts</p></li>
</ul><ul><li><p>sumDiffExp: Get summary table from diffExpAnalysis analysis</p></li>
</ul><ul><li><p>getDiffAnalysesSummary: ...</p></li>
</ul></div>
    <div id="plots">
    <h2>Plots</h2>
    



<ul><li><p>plotDiffAnalysis method draws a MAplot, a volcano plot and the 
p-values distribution from the results of a differential analysis.</p></li>
</ul><ul><li><p>plotHeatmapDesign method draws a heatmap from the results of a 
differential analysis.</p></li>
</ul><ul><li><p>plotBoxplotDE method draws a boxplot showing the expression of 
   given differentially expressed feature.</p></li>
</ul></div>
    <div id="references">
    <h2>References</h2>
    <p>Lambert, I., Paysant-Le Roux, C., Colella, S. et al. DiCoExpress: a tool 
to process multifactorial RNAseq experiments from quality controls to 
co-expression analysis through differential analysis based on contrasts 
inside GLM models. Plant Methods 16, 68 (2020).</p>
<p>Efstathiou G, Antonakis AN, Pavlopoulos GA, et al. ProteoSign: 
an end-user online differential proteomics statistical analysis platform. 
Nucleic Acids Res. 2017;45(W1):W300-W306.</p>
    </div>
    <div id="see-also">
    <h2>See also</h2>
    <div class="dont-index"><p><code>getDiffSettings</code>, <code>getDEList</code>,
 <code>getDEMatrix</code></p>
<p><code>plotDiffAnalysis</code>, <code>plotHeatmapDesign</code>,
<code>plotBoxplotDE</code></p></div>
    </div>
    <div id="author">
    <h2>Author</h2>
    <p>Christine Paysant-Le Roux, adapted by Nadia Bessoltane</p>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="co"># load ecoseed data</span></span></span>
<span class="r-in"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://forgemia.inra.fr/flomics/rflomics/" class="external-link">RFLOMICS</a></span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">ecoseed</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># create rflomicsMAE object with ecoseed data</span></span></span>
<span class="r-in"><span><span class="va">MAE</span> <span class="op">&lt;-</span> <span class="fu"><a href="createRflomicsMAE.html">createRflomicsMAE</a></span><span class="op">(</span></span></span>
<span class="r-in"><span>  projectName <span class="op">=</span> <span class="st">"Tests"</span>,</span></span>
<span class="r-in"><span>  omicsData   <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">ecoseed</span><span class="op">$</span><span class="va">RNAtest</span>, <span class="va">ecoseed</span><span class="op">$</span><span class="va">metatest</span>, <span class="va">ecoseed</span><span class="op">$</span><span class="va">protetest</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>  omicsNames  <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"RNAtest"</span>, <span class="st">"metatest"</span>, <span class="st">"protetest"</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>  omicsTypes  <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"RNAseq"</span>,<span class="st">"metabolomics"</span>,<span class="st">"proteomics"</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>  ExpDesign   <span class="op">=</span> <span class="va">ecoseed</span><span class="op">$</span><span class="va">design</span>,</span></span>
<span class="r-in"><span>  factorRef   <span class="op">=</span> <span class="va">ecoseed</span><span class="op">$</span><span class="va">factorRef</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Set the statistical model and contrasts to test</span></span></span>
<span class="r-in"><span><span class="va">formulae</span> <span class="op">&lt;-</span> <span class="fu"><a href="generateModelFormulae.html">generateModelFormulae</a></span><span class="op">(</span><span class="va">MAE</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">MAE</span> <span class="op">&lt;-</span> <span class="fu"><a href="generateModelFormulae.html">setModelFormula</a></span><span class="op">(</span><span class="va">MAE</span>, <span class="va">formulae</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">)</span>  </span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Get the contrasts List and choose the first 3 contrasts of type averaged</span></span></span>
<span class="r-in"><span><span class="va">contrastList</span> <span class="op">&lt;-</span> <span class="fu"><a href="generateExpressionContrast.html">generateExpressionContrast</a></span><span class="op">(</span><span class="va">MAE</span>, <span class="st">"averaged"</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">MAE</span> <span class="op">&lt;-</span> <span class="fu"><a href="generateExpressionContrast.html">setSelectedContrasts</a></span><span class="op">(</span><span class="va">MAE</span>, contrastList <span class="op">=</span> <span class="va">contrastList</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">2</span>, <span class="fl">3</span><span class="op">)</span>,<span class="op">]</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Run the data preprocessing and perform the differential analysis </span></span></span>
<span class="r-in"><span><span class="va">MAE</span> <span class="op">&lt;-</span> <span class="fu"><a href="runDataProcessing.html">runDataProcessing</a></span><span class="op">(</span><span class="va">MAE</span>, SE.name <span class="op">=</span> <span class="st">"protetest"</span>,  </span></span>
<span class="r-in"><span>                         transformMethod <span class="op">=</span> <span class="st">"log2"</span>,</span></span>
<span class="r-in"><span>                         normMethod <span class="op">=</span> <span class="st">"median"</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> [RFLOMICS] #    =&gt; select samples...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> [RFLOMICS] #    =&gt; transformation data...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> [RFLOMICS] #    =&gt; Run normalization...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> [RFLOMICS] #    =&gt; Compute PCA </span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">MAE</span> <span class="op">&lt;-</span> <span class="fu">runDiffAnalysis</span><span class="op">(</span><span class="va">MAE</span>, SE.name <span class="op">=</span> <span class="st">"protetest"</span>, </span></span>
<span class="r-in"><span>                       method <span class="op">=</span> <span class="st">"limmalmFit"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="co"># or</span></span></span>
<span class="r-in"><span><span class="co"># MAE[["protetest"]] &lt;- runDiffAnalysis(MAE[["protetest"]], </span></span></span>
<span class="r-in"><span><span class="co">#                                       method = "limmalmFit", </span></span></span>
<span class="r-in"><span><span class="co">#                                       contrastList = contrastList)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Filter the results of the differential analysis with new cut-off values </span></span></span>
<span class="r-in"><span><span class="co"># for p-value and fold change.</span></span></span>
<span class="r-in"><span><span class="va">MAE</span> <span class="op">&lt;-</span> <span class="fu">filterDiffAnalysis</span><span class="op">(</span><span class="va">MAE</span>, SE.name <span class="op">=</span> <span class="st">"protetest"</span>,</span></span>
<span class="r-in"><span>                          p.adj.cutoff <span class="op">=</span> <span class="fl">0.01</span>,</span></span>
<span class="r-in"><span>                          logFC.cutoff <span class="op">=</span> <span class="fl">0</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="co"># or</span></span></span>
<span class="r-in"><span><span class="co"># MAE[["protetest"]] &lt;- filterDiffAnalysis(MAE[["protetest"]], </span></span></span>
<span class="r-in"><span><span class="co">#                                          p.adj.cutoff = 0.01, </span></span></span>
<span class="r-in"><span><span class="co">#                                          logFC.cutoff = 0)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Access to the diff analysis settings</span></span></span>
<span class="r-in"><span><span class="co">## Get DE matrix from DiffExpAnalysis</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fu">getDEMatrix</span><span class="op">(</span><span class="va">MAE</span><span class="op">[[</span><span class="st">"protetest"</span><span class="op">]</span><span class="op">]</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>         DEF (temperatureMedium - temperatureLow) in mean</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1 AT2G42560                                            1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2 AT2G33070                                            1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3 AT4G27140                                            1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4 AT5G54740                                            1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 5 AT5G58070                                            1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 6 AT1G14940                                            1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   (temperatureElevated - temperatureLow) in mean</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1                                              1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2                                              1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3                                              1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4                                              1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 5                                              1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 6                                              1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   (temperatureElevated - temperatureMedium) in mean</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1                                                 0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2                                                 0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3                                                 0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4                                                 0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 5                                                 1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 6                                                 0</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">## Get union or intersection from list of contrasts</span></span></span>
<span class="r-in"><span><span class="fu">getDEList</span><span class="op">(</span><span class="va">MAE</span><span class="op">[[</span><span class="st">"protetest"</span><span class="op">]</span><span class="op">]</span>, contrasts <span class="op">=</span> <span class="st">"(temperatureMedium - temperatureLow) in mean"</span><span class="op">)</span> </span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [1] "AT2G42560" "AT2G33070" "AT4G27140" "AT5G54740" "AT5G58070" "AT1G14940"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [7] "AT1G77120" "AT4G04460" "AT1G20620" "AT2G38380" "AT3G21380" "AT5G57550"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [13] "AT2G21660" "AT1G14950" "AT1G70840" "AT3G60730" "AT1G17100" "AT4G15530"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [19] "AT3G54400" "AT5G52640" "AT2G28490" "AT2G23110" "AT2G24270" "AT1G14930"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [25] "AT2G46750" "AT5G66190" "AT3G14067" "AT3G12203" "AT3G09440" "AT5G26120"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [31] "AT3G44300" "AT5G12950" "AT1G65880" "AT2G41260" "AT5G35360" "AT1G43670"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [37] "AT5G55730" "AT4G34480" "AT1G18500" "AT3G23490" "AT5G43600" "AT5G34940"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [43] "AT5G04885" "AT1G52560" "AT4G19810" "AT2G43090" "AT1G20020" "AT1G26560"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [49] "AT4G35790" "AT3G07770" "AT4G16760"</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">## Get diff setting</span></span></span>
<span class="r-in"><span><span class="fu">getDiffSettings</span><span class="op">(</span><span class="va">MAE</span><span class="op">[[</span><span class="st">"protetest"</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $method</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "limmalmFit"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $p.adj.method</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "BH"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $p.adj.cutoff</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.01</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $abs.logFC.cutoff</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># generate plot results of a differential analysis</span></span></span>
<span class="r-in"><span><span class="va">thiscontrast</span> <span class="op">&lt;-</span> <span class="st">"(temperatureMedium - temperatureLow) in mean"</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">## generate MAplot from diff analysis</span></span></span>
<span class="r-in"><span><span class="co"># plotDiffAnalysis(MAE[["protetest"]], </span></span></span>
<span class="r-in"><span><span class="co">#                  contrastName = thiscontrast, </span></span></span>
<span class="r-in"><span><span class="co">#                  typeofplots = "MA.plot")</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">## plot the heatmap</span></span></span>
<span class="r-in"><span><span class="co"># plotHeatmapDesign(MAE[["protetest"]], </span></span></span>
<span class="r-in"><span><span class="co">#                   contrastName = thiscontrast)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">## plot boxplot with feature expression</span></span></span>
<span class="r-in"><span><span class="co"># plotBoxplotDE(MAE[["protetest"]], </span></span></span>
<span class="r-in"><span><span class="co">#               features = "AT1G47128", </span></span></span>
<span class="r-in"><span><span class="co">#               groupColor = "temperature")</span></span></span>
<span class="r-in"><span><span class="co"># plotBoxplotDE(MAE[["protetest"]], </span></span></span>
<span class="r-in"><span><span class="co">#               features = "AT1G79550", </span></span></span>
<span class="r-in"><span><span class="co">#               groupColor = "imbibition")</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># See runDiffAnalysis for an example that includes plotDiffAnalysis</span></span></span>
<span class="r-in"><span><span class="co"># See runDiffAnalysis for an example that includes plotHeatmapDesign</span></span></span>
<span class="r-in"><span><span class="co"># See runDiffAnalysis for an example that includes plotBoxplotDE</span></span></span>
<span class="r-in"><span><span class="co"># See runDiffAnalysis for an example that includes getDEMatrix</span></span></span>
<span class="r-in"><span><span class="co"># See runDiffAnalysis for an example that includes getDEList</span></span></span>
<span class="r-in"><span><span class="co"># See runDiffAnalysis for an example that includes getDiffSettings</span></span></span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Nadia Bessoltane, Delphine Charif, Audrey Hulot, Christine Paysant-Leroux, Gwendal Cueff.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.9.</p>
</div>

      </footer></div>

  


  

  </body></html>

