<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Data Exploratory and processing — runDataProcessing • RFLOMICS</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Data Exploratory and processing — runDataProcessing"><meta property="og:description" content='These functions applied a data processing (filtering, normalization 
and/or transformation, PCA) on RNAseq, proteomic, or metabolomic data.
runDataProcessing() calls the following functions:

filterLowAbundance:  This function aims at removing transcript from 
the count data matrix of an omic of type "RNAseq".
by applying filtering criterion described in reference.



runSampleFiltering: 
  This function applied sample filtering on an dataset.



runTransformData: 
   This function applied a transformation on dataset. The transformation 
method is chosen according to dataset omic type 
(RNaseq: none, metabolomics/proteomics: log2)



runNormalization: 
 This function applied a normalization on a dataset. 
The normalisation method is chosen according to dataset omic type 
(RNAseq: TMM, metabolics/proteomics: median)



runOmicsPCA: 
 This function performs a principal component analysis on omic 
data stored in an object of class RflomicsSE-class
Results are stored in the metadata slot of the same object. If a 
"Normalization" slot is present in the metadata slot, then data are 
normalized before running the PCA according to the indicated transform 
method.
This function performs a principal component analysis on omic 
data stored in an object of class RflomicsSE-class
Results are stored in the metadata slot of the same object. If a 
"Normalization" slot is present in the metadata slot, then data are 
normalized before running the PCA according to the indicated transform 
method.

checkExpDesignCompleteness: return a string with message.
   This method checks some experimental design characteristics.
   A complete design (all combinations of factor modalities with at 
   least 2 replicates for each have to be present) with
   at least one biological and one batch factors are required to use the 
   RFLOMICS workflow.



plotDataDistribution: return boxplot or density plot of expression
   or abundance distribution.

'><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">RFLOMICS</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.0.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../articles/RFLOMICS.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Manuals
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/RFLOMICS-cmd-line.html">RFLOMICS vignettes cmd line</a>
    </li>
    <li>
      <a href="../articles/RFLOMICS-input-data.html">RFLOMICS input data</a>
    </li>
  </ul></li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://forgemia.inra.fr/flomics/rflomics" class="external-link">
    <span class="fa fa-gitlab"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Data Exploratory and processing</h1>
    
    <div class="hidden name"><code>runDataProcessing.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>These functions applied a data processing (filtering, normalization 
and/or transformation, PCA) on RNAseq, proteomic, or metabolomic data.</p>
<p>runDataProcessing() calls the following functions:</p>
<ul><li><p>filterLowAbundance:  This function aims at removing transcript from 
the count data matrix of an omic of type "RNAseq".
by applying filtering criterion described in reference.</p></li>
</ul><ul><li><p>runSampleFiltering: 
  This function applied sample filtering on an dataset.</p></li>
</ul><ul><li><p>runTransformData: 
   This function applied a transformation on dataset. The transformation 
method is chosen according to dataset omic type 
(RNaseq: none, metabolomics/proteomics: log2)</p></li>
</ul><ul><li><p>runNormalization: 
 This function applied a normalization on a dataset. 
The normalisation method is chosen according to dataset omic type 
(RNAseq: TMM, metabolics/proteomics: median)</p></li>
</ul><ul><li><p>runOmicsPCA: 
 This function performs a principal component analysis on omic 
data stored in an object of class <a href="RflomicsSE-class.html">RflomicsSE-class</a>
Results are stored in the metadata slot of the same object. If a 
"Normalization" slot is present in the metadata slot, then data are 
normalized before running the PCA according to the indicated transform 
method.</p></li>
</ul><p>This function performs a principal component analysis on omic 
data stored in an object of class <a href="RflomicsSE-class.html">RflomicsSE-class</a>
Results are stored in the metadata slot of the same object. If a 
"Normalization" slot is present in the metadata slot, then data are 
normalized before running the PCA according to the indicated transform 
method.</p>
<ul><li><p>checkExpDesignCompleteness: return a string with message.
   This method checks some experimental design characteristics.
   A complete design (all combinations of factor modalities with at 
   least 2 replicates for each have to be present) with
   at least one biological and one batch factors are required to use the 
   RFLOMICS workflow.</p></li>
</ul><ul><li><p>plotDataDistribution: return boxplot or density plot of expression
   or abundance distribution.</p></li>
</ul></div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="co"># S4 method for RflomicsSE</span></span>
<span><span class="fu">runDataProcessing</span><span class="op">(</span></span>
<span>  <span class="va">object</span>,</span>
<span>  samples <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  lowCountFiltering_strategy <span class="op">=</span> <span class="st">"NbReplicates"</span>,</span>
<span>  lowCountFiltering_CPM_Cutoff <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  normMethod <span class="op">=</span> <span class="st">"none"</span>,</span>
<span>  transformMethod <span class="op">=</span> <span class="st">"none"</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># S4 method for RflomicsMAE</span></span>
<span><span class="fu">runDataProcessing</span><span class="op">(</span></span>
<span>  <span class="va">object</span>,</span>
<span>  samples <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  lowCountFiltering_strategy <span class="op">=</span> <span class="st">"NbReplicates"</span>,</span>
<span>  lowCountFiltering_CPM_Cutoff <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  normMethod <span class="op">=</span> <span class="st">"none"</span>,</span>
<span>  transformMethod <span class="op">=</span> <span class="st">"none"</span>,</span>
<span>  <span class="va">SE.name</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># S4 method for RflomicsSE</span></span>
<span><span class="fu">filterLowAbundance</span><span class="op">(</span></span>
<span>  <span class="va">object</span>,</span>
<span>  filterMethod <span class="op">=</span> <span class="st">"CPM"</span>,</span>
<span>  filterStrategy <span class="op">=</span> <span class="st">"NbConditions"</span>,</span>
<span>  cpmCutoff <span class="op">=</span> <span class="fl">5</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># S4 method for RflomicsMAE</span></span>
<span><span class="fu">filterLowAbundance</span><span class="op">(</span></span>
<span>  <span class="va">object</span>,</span>
<span>  <span class="va">SE.name</span>,</span>
<span>  filterMethod <span class="op">=</span> <span class="st">"CPM"</span>,</span>
<span>  filterStrategy <span class="op">=</span> <span class="st">"NbConditions"</span>,</span>
<span>  cpmCutoff <span class="op">=</span> <span class="fl">5</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># S4 method for RflomicsSE</span></span>
<span><span class="fu">runSampleFiltering</span><span class="op">(</span><span class="va">object</span>, samples <span class="op">=</span> <span class="cn">NULL</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># S4 method for RflomicsMAE</span></span>
<span><span class="fu">runSampleFiltering</span><span class="op">(</span><span class="va">object</span>, <span class="va">SE.name</span>, samples <span class="op">=</span> <span class="cn">NULL</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># S4 method for RflomicsSE</span></span>
<span><span class="fu">runTransformData</span><span class="op">(</span><span class="va">object</span>, transformMethod <span class="op">=</span> <span class="cn">NULL</span>, modifyAssay <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># S4 method for RflomicsMAE</span></span>
<span><span class="fu">runTransformData</span><span class="op">(</span><span class="va">object</span>, <span class="va">SE.name</span>, transformMethod <span class="op">=</span> <span class="cn">NULL</span>, modifyAssay <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># S4 method for RflomicsSE</span></span>
<span><span class="fu">runNormalization</span><span class="op">(</span><span class="va">object</span>, normMethod <span class="op">=</span> <span class="cn">NULL</span>, modifyAssay <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># S4 method for RflomicsMAE</span></span>
<span><span class="fu">runNormalization</span><span class="op">(</span><span class="va">object</span>, <span class="va">SE.name</span>, normMethod <span class="op">=</span> <span class="cn">NULL</span>, modifyAssay <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># S4 method for RflomicsSE</span></span>
<span><span class="fu">runOmicsPCA</span><span class="op">(</span><span class="va">object</span>, ncomp <span class="op">=</span> <span class="fl">5</span>, raw <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># S4 method for RflomicsMAE</span></span>
<span><span class="fu">runOmicsPCA</span><span class="op">(</span><span class="va">object</span>, <span class="va">SE.name</span>, ncomp <span class="op">=</span> <span class="fl">5</span>, raw <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># S4 method for RflomicsSE</span></span>
<span><span class="fu">checkExpDesignCompleteness</span><span class="op">(</span><span class="va">object</span>, sampleList <span class="op">=</span> <span class="cn">NULL</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># S4 method for RflomicsMAE</span></span>
<span><span class="fu">checkExpDesignCompleteness</span><span class="op">(</span><span class="va">object</span>, <span class="va">omicName</span>, sampleList <span class="op">=</span> <span class="cn">NULL</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># S4 method for RflomicsSE</span></span>
<span><span class="fu">getProcessedData</span><span class="op">(</span><span class="va">object</span>, NormTrans <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># S4 method for RflomicsMAE</span></span>
<span><span class="fu">getProcessedData</span><span class="op">(</span><span class="va">object</span>, <span class="va">SE.name</span>, NormTrans <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># S4 method for RflomicsSE</span></span>
<span><span class="fu">getTransSettings</span><span class="op">(</span><span class="va">object</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># S4 method for RflomicsMAE</span></span>
<span><span class="fu">getTransSettings</span><span class="op">(</span><span class="va">object</span>, <span class="va">SE.name</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># S4 method for RflomicsSE</span></span>
<span><span class="fu">getFilterSettings</span><span class="op">(</span><span class="va">object</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># S4 method for RflomicsMAE</span></span>
<span><span class="fu">getFilterSettings</span><span class="op">(</span><span class="va">object</span>, <span class="va">SE.name</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># S4 method for RflomicsSE</span></span>
<span><span class="fu">getFilteredFeatures</span><span class="op">(</span><span class="va">object</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># S4 method for RflomicsMAE</span></span>
<span><span class="fu">getFilteredFeatures</span><span class="op">(</span><span class="va">object</span>, <span class="va">SE.name</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># S4 method for RflomicsSE</span></span>
<span><span class="fu">getCoeffNorm</span><span class="op">(</span><span class="va">object</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># S4 method for RflomicsMAE</span></span>
<span><span class="fu">getCoeffNorm</span><span class="op">(</span><span class="va">object</span>, <span class="va">SE.name</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># S4 method for RflomicsSE</span></span>
<span><span class="fu">getNormSettings</span><span class="op">(</span><span class="va">object</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># S4 method for RflomicsMAE</span></span>
<span><span class="fu">getNormSettings</span><span class="op">(</span><span class="va">object</span>, <span class="va">SE.name</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># S4 method for RflomicsSE</span></span>
<span><span class="fu">plotLibrarySize</span><span class="op">(</span><span class="va">object</span>, raw <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># S4 method for RflomicsMAE</span></span>
<span><span class="fu">plotLibrarySize</span><span class="op">(</span><span class="va">object</span>, <span class="va">SE.name</span>, raw <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># S4 method for RflomicsSE</span></span>
<span><span class="fu">plotDataDistribution</span><span class="op">(</span><span class="va">object</span>, plot <span class="op">=</span> <span class="st">"boxplot"</span>, raw <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># S4 method for RflomicsMAE</span></span>
<span><span class="fu">plotDataDistribution</span><span class="op">(</span><span class="va">object</span>, <span class="va">SE.name</span>, plot <span class="op">=</span> <span class="st">"boxplot"</span>, raw <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># S4 method for RflomicsSE</span></span>
<span><span class="fu">plotOmicsPCA</span><span class="op">(</span></span>
<span>  <span class="va">object</span>,</span>
<span>  raw <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"raw"</span>, <span class="st">"norm"</span><span class="op">)</span>,</span>
<span>  axes <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">2</span><span class="op">)</span>,</span>
<span>  groupColor <span class="op">=</span> <span class="st">"groups"</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># S4 method for RflomicsMAE</span></span>
<span><span class="fu">plotOmicsPCA</span><span class="op">(</span><span class="va">object</span>, <span class="va">SE.name</span>, <span class="va">raw</span>, axes <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">2</span><span class="op">)</span>, groupColor <span class="op">=</span> <span class="st">"groups"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># S4 method for RflomicsSE</span></span>
<span><span class="fu">plotExpDesignCompleteness</span><span class="op">(</span><span class="va">object</span>, sampleList <span class="op">=</span> <span class="cn">NULL</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># S4 method for RflomicsMAE</span></span>
<span><span class="fu">plotExpDesignCompleteness</span><span class="op">(</span><span class="va">object</span>, <span class="va">omicName</span>, sampleList <span class="op">=</span> <span class="cn">NULL</span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>object</dt>
<dd><p>An object of class <a href="RflomicsSE-class.html">RflomicsSE-class</a>.</p></dd>


<dt>samples</dt>
<dd><p>samples to keep.</p></dd>


<dt>lowCountFiltering_strategy</dt>
<dd><p>strategy of RNAseq low count filtering. 
Mandatory for RNAseq data. Default value: "NbReplicates".</p></dd>


<dt>lowCountFiltering_CPM_Cutoff</dt>
<dd><p>CPM cutoff for RNAseq low count filtering.
Mandatory for RNAseq data. Default value: 1.</p></dd>


<dt>normMethod</dt>
<dd><p>Normalization method</p></dd>


<dt>transformMethod</dt>
<dd><p>The transformation to store in the metadata or to 
store and apply if modifyAssay is TRUE.</p></dd>


<dt>SE.name</dt>
<dd><p>the name of the data the normalization have to be applied to.</p></dd>


<dt>filterMethod</dt>
<dd><p>The filtering model ("CPM")</p></dd>


<dt>filterStrategy</dt>
<dd><p>The filtering strategy 
("NbConditions" or "NbReplicates")</p></dd>


<dt>cpmCutoff</dt>
<dd><p>The CPM cutoff.</p></dd>


<dt>modifyAssay</dt>
<dd><p>Does the normalization have to be applied or just 
stored for later? Recommended it stays FALSE.</p></dd>


<dt>ncomp</dt>
<dd><p>Number of components to compute. Default is 5.</p></dd>


<dt>raw</dt>
<dd><p>This argument indicates whether the scaled PCA has to be 
performed on raw [‘raw’] or normalized [‘norm’] data.</p></dd>


<dt>sampleList</dt>
<dd><p>list of samples to check.</p></dd>


<dt>omicName</dt>
<dd><p>a character string with the name of the dataset</p></dd>


<dt>NormTrans</dt>
<dd><p>boolean. If TRUE, returned Data are normalized 
and transformed.</p></dd>


<dt>plot</dt>
<dd><p>plot type ("boxplot" or "density")</p></dd>


<dt>axes</dt>
<dd><p>A vector giving the two axis that have to be drawn for the 
factorial map</p></dd>


<dt>groupColor</dt>
<dd><p>All combination of level's factor</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    

<p>An object of class <a href="RflomicsSE-class.html">RflomicsSE</a> or class <a href="RflomicsSE-class.html">RflomicsSE</a></p>



<p>An object of class <a href="RflomicsSE-class.html">RflomicsSE</a></p>


<p>The applied normalization method and computed scaling factors 
(by samples) are stored as a named list
("normalization") of two elements (respectively "method" and 
"coefNorm") in the metadata slot of a
given data set, stored itself in the ExperimentList slot of a 
<a href="RflomicsSE-class.html">RflomicsSE</a> object.</p>


<p>An object of class <a href="RflomicsSE-class.html">RflomicsSE</a></p>


    </div>
    <div id="details">
    <h2>Details</h2>
    <p>filterLowAbundance(): By default, gene/transcript with 0 count 
are removed from the data. The function then
computes the count per million or read (CPM) for each gene 
in each sample and gives by
genes the number of sample(s) which are over the cpmCutoff 
(NbOfsample_over_cpm).
Then Two filtering strategies are proposed:</p><ul><li><p>NbConditions:  keep gene if the NbOfsample_over_cpm &gt;= NbConditions</p></li>
<li><p>NbReplicates:  keep gene if the NbOfsample_over_cpm &gt;= min(NbReplicat)</p></li>
<li><p>filterByExpr: the default filtering method implemented 
in the edgeR filterByExpr() function.</p></li>
</ul></div>
    <div id="accessors">
    <h2>Accessors</h2>
    


 
<ul><li><p>getProcessedData: return a processed data matrix  
   (filtering, normalization and/or transformation)</p></li>
</ul><ul><li><p>getTransSettings: return a list of transformation settings 
   of a given omic dataset</p></li>
</ul><ul><li><p>getFilterSettings: return a list the filtering settings of a given 
   omic dataset</p></li>
</ul><ul><li><p>getFilteredFeatures: return a vector of filtered features of a given 
   omic dataset</p></li>
</ul><ul><li><p>getCoeffNorm: return a named vector with normalization coefficients 
   of a given omic dataset</p></li>
</ul><ul><li><p>getNormSettings: return a list of normalization settings 
   of a given omic dataset</p></li>
</ul></div>
    <div id="plots">
    <h2>Plots</h2>
    


 
<ul><li><p>plotLibrarySize: return barplot of library size by sample.</p></li>
</ul><ul><li><p>plotOmicsPCA: 
This function plot the factorial map from a PCA object stored
in a <a href="RflomicsSE-class.html">RflomicsSE-class</a> object. By default, samples are
colored by groups (all combinations of level's factor)</p></li>
</ul><ul><li><p>plotExpDesignCompleteness: 
This method checks that experimental design constraints are satisfied and 
plot a summary of the design.
A complete design (all combinations of factor modalities with at least 2 
replicates for each have to be present) 
with at least one biological and one batch factors are required to use the 
RFLOMICS workflow.</p></li>
</ul></div>
    <div id="references">
    <h2>References</h2>
    <p>Lambert, I., Paysant-Le Roux, C., Colella, S. et al. 
DiCoExpress: a tool to process multifactorial
RNAseq experiments from quality controls to co-expression analysis through 
differential analysis based on contrasts inside GLM models. 
Plant Methods 16, 68 (2020).</p>
    </div>
    <div id="see-also">
    <h2>See also</h2>
    <div class="dont-index"><p><a href="RflomicsMAE-class.html">RflomicsMAE-class</a> 
<a href="RflomicsSE-class.html">RflomicsSE-class</a> 
getProcessedData 
getTransSettings 
getFilterSettings
getFilteredFeatures
getCoeffNorm
getNormSettings
plotLibrarySize
plotDataDistribution
plotOmicsPCA</p>
<p>edgeR::filterByExpr</p></div>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="co"># load ecoseed data</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">ecoseed</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># create rflomicsMAE object with ecoseed data</span></span></span>
<span class="r-in"><span><span class="va">MAE</span> <span class="op">&lt;-</span> <span class="fu">RFLOMICS</span><span class="fu">::</span><span class="fu"><a href="createRflomicsMAE.html">createRflomicsMAE</a></span><span class="op">(</span></span></span>
<span class="r-in"><span>  projectName <span class="op">=</span> <span class="st">"Tests"</span>,</span></span>
<span class="r-in"><span>  omicsData   <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">ecoseed</span><span class="op">$</span><span class="va">RNAtest</span>, <span class="va">ecoseed</span><span class="op">$</span><span class="va">metatest</span>, <span class="va">ecoseed</span><span class="op">$</span><span class="va">protetest</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>  omicsNames  <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"RNAtest"</span>, <span class="st">"metatest"</span>, <span class="st">"protetest"</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>  omicsTypes  <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"RNAseq"</span>,<span class="st">"metabolomics"</span>,<span class="st">"proteomics"</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>  ExpDesign   <span class="op">=</span> <span class="va">ecoseed</span><span class="op">$</span><span class="va">design</span>,</span></span>
<span class="r-in"><span>  factorRef   <span class="op">=</span> <span class="va">ecoseed</span><span class="op">$</span><span class="va">factorRef</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Set the statistical model</span></span></span>
<span class="r-in"><span><span class="va">formulae</span> <span class="op">&lt;-</span> <span class="fu"><a href="generateModelFormulae.html">generateModelFormulae</a></span><span class="op">(</span><span class="va">MAE</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">MAE</span> <span class="op">&lt;-</span> <span class="fu"><a href="generateModelFormulae.html">setModelFormula</a></span><span class="op">(</span><span class="va">MAE</span>, <span class="va">formulae</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">)</span>  </span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># set the contrast list</span></span></span>
<span class="r-in"><span><span class="va">contrastList</span> <span class="op">&lt;-</span> <span class="fu"><a href="generateExpressionContrast.html">generateExpressionContrast</a></span><span class="op">(</span><span class="va">MAE</span>, <span class="st">"averaged"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">MAE</span> <span class="op">&lt;-</span> <span class="fu"><a href="generateExpressionContrast.html">setSelectedContrasts</a></span><span class="op">(</span><span class="va">MAE</span>, contrastList <span class="op">=</span> <span class="va">contrastList</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">2</span>, <span class="fl">3</span><span class="op">)</span>,<span class="op">]</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Data processing of RNAseq dataset : RNAtest</span></span></span>
<span class="r-in"><span><span class="co">## using data processing functions for RNAseq data</span></span></span>
<span class="r-in"><span><span class="co">### filter low RNAseq count </span></span></span>
<span class="r-in"><span><span class="co"># MAE &lt;- filterLowAbundance(MAE, SE.name = "RNAtest",</span></span></span>
<span class="r-in"><span><span class="co">#                           filterStrategy = "NbReplicates", </span></span></span>
<span class="r-in"><span><span class="co">#                           cpmCutoff = 1)</span></span></span>
<span class="r-in"><span><span class="co"># ### filter outlier samples</span></span></span>
<span class="r-in"><span><span class="co"># MAE &lt;- runSampleFiltering(MAE, SE.name = "RNAtest", </span></span></span>
<span class="r-in"><span><span class="co">#                           samples = colnames(MAE[["RNAtest"]])[-1])</span></span></span>
<span class="r-in"><span><span class="co"># ### data normalisation outlier samples</span></span></span>
<span class="r-in"><span><span class="co"># MAE &lt;- runNormalization(MAE, SE.name = "RNAtest", </span></span></span>
<span class="r-in"><span><span class="co">#                         normMethod = "TMM")</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">## use runDataProcessing function that combines the previous three functions</span></span></span>
<span class="r-in"><span><span class="va">MAE</span> <span class="op">&lt;-</span> <span class="fu">runDataProcessing</span><span class="op">(</span><span class="va">MAE</span>, SE.name <span class="op">=</span> <span class="st">"RNAtest"</span>, </span></span>
<span class="r-in"><span>                         samples <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">MAE</span><span class="op">[[</span><span class="st">"RNAtest"</span><span class="op">]</span><span class="op">]</span><span class="op">)</span><span class="op">[</span><span class="op">-</span><span class="fl">1</span><span class="op">]</span>, </span></span>
<span class="r-in"><span>                         lowCountFiltering_strategy <span class="op">=</span> <span class="st">"NbReplicates"</span>, </span></span>
<span class="r-in"><span>                         lowCountFiltering_CPM_Cutoff <span class="op">=</span> <span class="fl">1</span>, </span></span>
<span class="r-in"><span>                         normMethod <span class="op">=</span> <span class="st">"TMM"</span><span class="op">)</span> </span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> [RFLOMICS] #    =&gt; select samples...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> [RFLOMICS] #    =&gt; Low counts Filtering...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> [RFLOMICS] #    =&gt; Counts normalization...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> [RFLOMICS] #    =&gt; Compute PCA </span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">## check completness of RNAtest data</span></span></span>
<span class="r-in"><span><span class="fu">checkExpDesignCompleteness</span><span class="op">(</span><span class="va">MAE</span>, omicName <span class="op">=</span> <span class="st">"RNAtest"</span><span class="op">)</span><span class="op">$</span><span class="va">messages</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "The experimental design is complete and balanced."</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Data processing of proteimics dataset : protetest</span></span></span>
<span class="r-in"><span><span class="co"># ## transform data</span></span></span>
<span class="r-in"><span><span class="co"># MAE &lt;- runTransformData(MAE, SE.name = "protetest",  transformMethod = "log2")</span></span></span>
<span class="r-in"><span><span class="co"># ## normalise data</span></span></span>
<span class="r-in"><span><span class="co"># MAE &lt;- runNormalization(MAE, SE.name = "protetest", normMethod = "median") </span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">## use runDataProcessing function</span></span></span>
<span class="r-in"><span><span class="va">MAE</span> <span class="op">&lt;-</span> <span class="fu">runDataProcessing</span><span class="op">(</span><span class="va">MAE</span>, SE.name <span class="op">=</span> <span class="st">"protetest"</span>, </span></span>
<span class="r-in"><span>                         normMethod <span class="op">=</span> <span class="st">"median"</span>, </span></span>
<span class="r-in"><span>                         transformMethod <span class="op">=</span> <span class="st">"log2"</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> [RFLOMICS] #    =&gt; select samples...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> [RFLOMICS] #    =&gt; transformation data...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> [RFLOMICS] #    =&gt; Run normalization...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> [RFLOMICS] #    =&gt; Compute PCA </span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># plotExpDesignCompleteness(MAE[["RNAtest"]])</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># plot Library Size</span></span></span>
<span class="r-in"><span><span class="co"># plotDataDistribution(MAE[["RNAtest"]], raw=TRUE)</span></span></span>
<span class="r-in"><span><span class="co"># plotDataDistribution(MAE[["RNAtest"]], raw=FALSE)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># plot gene expression distribution</span></span></span>
<span class="r-in"><span><span class="co"># plotDataDistribution(MAE[["RNAtest"]], raw=TRUE, plot = "boxplot")</span></span></span>
<span class="r-in"><span><span class="co"># plotDataDistribution(MAE[["RNAtest"]], raw=FALSE, plot = "boxplot")</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># plot PCA </span></span></span>
<span class="r-in"><span><span class="co"># plotOmicsPCA(MAE[["RNAtest"]], raw="raw", groupColor = "imbibition")</span></span></span>
<span class="r-in"><span><span class="co"># plotOmicsPCA(MAE[["RNAtest"]], raw="norm", groupColor = "imbibition")</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># See runDataProcessing for an example that includes getProcessedData</span></span></span>
<span class="r-in"><span><span class="co"># See runDataProcessing for an example that includes getTransSettings</span></span></span>
<span class="r-in"><span><span class="co"># See runDataProcessing for an example that includes getFilterSettings</span></span></span>
<span class="r-in"><span><span class="co"># See runDataProcessing for an example that includes getFilteredFeatures</span></span></span>
<span class="r-in"><span><span class="co"># See runDataProcessing for an example that includes getCoeffNorm</span></span></span>
<span class="r-in"><span><span class="co"># See runDataProcessing for an example that includes getNormSettings</span></span></span>
<span class="r-in"><span><span class="co"># See runDataProcessing for an example that includes plotLibrarySize</span></span></span>
<span class="r-in"><span><span class="co"># See runDataProcessing for an example that includes plotDataDistribution</span></span></span>
<span class="r-in"><span><span class="co"># See runDataProcessing for an example that includes plotOmicsPCA</span></span></span>
<span class="r-in"><span><span class="co"># See runDataProcessing for an example that includes plotExpDesignCompleteness</span></span></span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Nadia Bessoltane, Delphine Charif, Audrey Hulot, Christine Paysant-Leroux, Gwendal Cueff.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.9.</p>
</div>

      </footer></div>

  


  

  </body></html>

